---
title: "Spatial vs. temporal storage: The dual role of microbial seed
banks in driving geographical patterns of microbial diversity"

geometry: margin=2.54cm
date: "August 4, 2015"
output: pdf_document
header-includes: \usepackage{array}
---

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Retrieve and set the working directory
rm(list=ls())
getwd()
setwd("~/GitHub/Dimensions/Aim3")
```

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load packages
require("sp")          # Classes and methods for handling spatial data
require("geoR")        # Methods for geostatistical analyses
require("rgdal")       # Geospatial Data Abstraction Library
require("raster")      # Methods to create a RasterLayer object
require("maptools")    # Tools for manipulating and reading geospatial data

require("picante")
require("ape")
require("seqinr")
require("vegan") # biodiversity estimators and related functions
require("fossil")
require("simba")
require("reshape")

require(ggplot2)
require(reshape)
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Import functions
# import code file that includes a function for reading in output files 
#from community sequencing software `mothur` (http://www.mothur.org/).
source("~/GitHub/Dimensions/Aim3/bin/MothurTools.R")
# import code for constructing species-area relationships by aggregating
# sites at random, with or without respect to distance
source("~/GitHub/Dimensions/Aim3/bin/SAR.R")
# import code for calculating diversity metrics
source("~/GitHub/Dimensions/Aim3/bin/metrics.R")
load(file = "~/GitHub/Dimensions/Aim3/Mothur/INPond_Initial.RData")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load Environmental and Geographical Data
env <- read.table("~/GitHub/Dimensions/Aim3/DATA/EnvData/20130801_PondDataMod.csv",
                  sep = ",", header = TRUE)
lats <- as.numeric(env[, 3]) # latitudes (north and south)
lons <- as.numeric(env[, 4]) # longitudes (east and west)

# Geographic Distances (Kilometers) Among Ponds
long.lat <- as.matrix(cbind(env$long, env$lat))
coord.dist <- earth.dist(long.lat, dist = TRUE)
coord.dist <- log(coord.dist)
coord.dist[which(!is.finite(coord.dist))] = NA
coord.dist.ls <- liste(coord.dist, entry = "geo.dist")

# Select DNA Data: Use the `grep()` Command and Rename with `gsub()`
# The active portion, based on cDNA 
active.comm <- Pond97[grep("*-cDNA", rownames(Pond97)), ]
rownames(active.comm) <- gsub("\\-cDNA", "", rownames(active.comm))
rownames(active.comm) <- gsub("\\_", "", rownames(active.comm))

# The community without respect to active or not, 16S rRNA gene sequences
all.comm <- Pond97[grep("*-DNA", rownames(Pond97)), ]
rownames(all.comm) <- gsub("\\-DNA", "", rownames(all.comm))
rownames(all.comm) <- gsub("\\_", "", rownames(all.comm))

# Remove Sites Not in the Environmental Data Set
active.comm <- active.comm[rownames(active.comm)  %in% env$Sample_ID, ]
all.comm <- all.comm[rownames(all.comm)  %in% env$Sample_ID, ]

# Remove Zero-Occurrence Taxa 
active.comm <- active.comm[ , colSums(active.comm) > 0]
all.comm <- all.comm[ , colSums(all.comm) > 0]

# Import Taxonomy Data Using `read.tax()` from Source Code
# tax <- read.tax(taxonomy = "~/GitHub/Dimensions/Aim3/Mothur/INPonds.bac.final.0.03.taxonomy")

# N equals numbers of reads
env$active.N <- as.vector(rowSums(active.comm)) 
env$all.N <- as.vector(rowSums(all.comm))

# S equals the number of non-zero abundances
env$active.S <- S.obs(active.comm)
env$all.S <- S.obs(all.comm)

# Diversity is Shannon's
env$active.H <- as.vector(diversity(active.comm, index = "shannon"))
env$all.H <- as.vector(diversity(all.comm, index = "shannon"))

# Evenness is Simpsons; divide Simpson's Diversity by S
env$active.De <- as.vector(diversity(active.comm, index = "invsimpson")/env$active.S)
env$all.De <- as.vector(diversity(all.comm, index = "invsimpson")/env$all.S)
```


## 1.) Dependence of the TAR on the number (detection) of rare taxa
The species-area relationshp describes the rate at which species are 
discovred with increasing area. The SAR one of ecology's oldest and most 
intensively studied patterns. Arrhenius (1921) first described the general 
form of the *species-area relationship (SAR)* as a power-law: $S = cA^{z}$ 
where S is species richnness and A is area. Arrhenius's formula predicts a 
rate of increase in richness that is approximately linear in log-log space.
That is, $log(S) = c + zlog(A)$, where z is the scaling exponent.

```{r, results='hide', echo=F, message=F, warning=F, fig.width=7, fig.height=4}
par(mfrow=c(1, 1))
#Taxa-area relationship
iterations = 4
pond.areas <- as.vector(pi * (env$Diameter/2)^2) # Find areas of all 51 ponds
cutoffs <- c(15, 30, 60, 125, 250)#, 500, 1000, 2000, 4000, 8000, 16000, 22098)

active.zs.dist <- c()
all.zs.dist <- c()
active.zs.rand <- c()
all.zs.rand <- c()

for (cutoff in cutoffs){
  print(cutoff)
  all.sub.comm <- all.comm[, 1:cutoff]
  active.sub.comm <- active.comm[, 1:cutoff]

  # all, distance
  sar <- as.data.frame(SAR.accum.dist(all.sub.comm, 
                                      coord.dist.ls$geo.dist, iterations))
  OLS <- lm(sar$V2 ~ sar$V1)
  all.zs.dist <- c(all.zs.dist, coefficients(OLS)[2])
  
  # active, distance
  sar <- as.data.frame(SAR.accum.dist(active.sub.comm, 
                                      coord.dist.ls$geo.dist, iterations))
  OLS <- lm(sar$V2 ~ sar$V1)
  active.zs.dist <- c(active.zs.dist, coefficients(OLS)[2])
  
  # all, random
  sar <- as.data.frame(SAR.rand.accum(all.sub.comm, iterations))
  OLS <- lm(sar$V2 ~ sar$V1)
  all.zs.rand <- c(all.zs.rand, coefficients(OLS)[2])
  
  # active, distance
  sar <- as.data.frame(SAR.rand.accum(active.sub.comm, iterations))
  OLS <- lm(sar$V2 ~ sar$V1)
  active.zs.rand <- c(active.zs.rand, coefficients(OLS)[2])
  }

z.lists <- as.data.frame(c(all.zs.dist, active.zs.dist, 
                           all.zs.rand, active.zs.rand))
  

z.df <- data.frame(cutoff = cutoffs,
                 All_Distance = all.zs.dist, 
                 Active_Distance = active.zs.dist, 
                 All_Random = all.zs.rand, 
                 Active_Random = active.zs.rand)
  
z.df <- melt(z.df,  id.vars = 'cutoff', variable_name = 'series')

# plot on same grid, each series colored differently
ggplot(z.df, aes(cutoff, value)) + 
  geom_line(aes(colour = series)) +
  labs(x= "Number of OTU's", y="Slope of the TAR (z-value)") +
  labs(title="Dependence of TAR slope on rare taxa") +
  theme(legend.position="top") +
  guides(fill=guide_legend(title=NULL)) +
  theme(legend.title=element_blank())
```


## 3.) Slight differences in distance-decay relationships

```{r, results='hide', echo=F, message=F, warning=F, fig.width=7, fig.height=4}
par(mfrow=c(1, 1))

# Geographic Distances (Kilometers) Among Ponds
long.lat <- as.matrix(cbind(env$long, env$lat))
coord.dist <- earth.dist(long.lat, dist = TRUE)
coord.dist <- log(coord.dist)
coord.dist[which(!is.finite(coord.dist))] = NA
coord.dist.ls <- liste(coord.dist, entry = "geo.dist")

cutoffs <- c(15, 30, 60, 125, 250, 500, 1000, 2000, 4000, 8000, 16000, 22098)

active.dd.BC <- c()
all.dd.BC <- c()

active.dd.S <- c()
all.dd.S <- c()

for (cutoff in cutoffs){
  print(cutoff)
  all.sub.comm <- all.comm[, 1:cutoff]
  active.sub.comm <- active.comm[, 1:cutoff]

  # Taxonomic Distances Among Ponds (Bray-Curits)
  active.bray.curtis.dist <- 1 - vegdist(active.sub.comm,
                                         method="bray", binary=FALSE) 
  active.bray.curtis.dist <- log(active.bray.curtis.dist)
  all.bray.curtis.dist <- 1 - vegdist(all.sub.comm,
                                      method="bray", binary=FALSE)
  all.bray.curtis.dist <- log(all.bray.curtis.dist)

  active.sorensen.dist <- 1 - vegdist(active.sub.comm,
                                      method="bray", binary=TRUE) 
  active.sorensen.dist <- log(active.sorensen.dist)
  all.sorensen.dist <- 1 - vegdist(all.sub.comm,
                                   method="bray", binary=TRUE)
  all.sorensen.dist <- log(all.sorensen.dist)

  # Transform All Distances Into List Format:
  active.bray.curtis.dist.ls <- liste(active.bray.curtis.dist, entry = "bray.curtis")
  all.bray.curtis.dist.ls <- liste(all.bray.curtis.dist, entry = "bray.curtis")
  active.sorensen.dist.ls <- liste(active.sorensen.dist, entry = "sorensen")
  all.sorensen.dist.ls <- liste(all.sorensen.dist, entry = "sorensen")

  # Create a Data Frame from the Lists of Distances
  df <- data.frame(coord.dist.ls, 
                   active.bray.curtis.dist.ls[, 3], 
                   all.bray.curtis.dist.ls[, 3], 
                   active.sorensen.dist.ls[, 3], 
                   all.sorensen.dist.ls[, 3])
          
  names(df)[4:7] <- c("active.bray.curtis", "all.bray.curtis",
                      "active.sorensen", "all.sorensen")
  attach(df)
  
  # Regression for active Bray Curtis
  DD.active.bray.curtis <- lm(active.bray.curtis.dist ~ geo.dist)
  slope <- round(coefficients(DD.active.bray.curtis)[2], 3)
  active.dd.BC <- c(active.dd.BC, slope)
  #p <- round(summary(DD.all.bray.curtis)$coefficients[8], 3)
    
  # Regression for all Bray Curtis 
  DD.all.bray.curtis <- lm(all.bray.curtis.dist ~ geo.dist)
  slope <- round(coefficients(DD.all.bray.curtis)[2], 3)
  all.dd.BC <- c(all.dd.BC, slope)

  # Regression for active Sorensen
  DD.active.sorensen <- lm(active.sorensen.dist ~ geo.dist)
  slope <- round(coefficients(DD.active.sorensen)[2], 3)
  active.dd.S <- c(active.dd.S, slope)

  # Regression for all Sorensen 
  DD.all.sorensen <- lm(all.sorensen.dist ~ geo.dist)
  slope <- round(coefficients(DD.all.sorensen)[2], 3)
  all.dd.S <- c(all.dd.S, slope)
  }
  
# Set Initial Plot Parameters
par(mfrow=c(2, 2))

dd.df <- data.frame(cutoff = cutoffs,
                 All_BrayCurtis = all.dd.BC, 
                 Active_BrayCurtis = active.dd.BC, 
                 All_Sorensen = all.dd.S, 
                 Active_Sorensen = active.dd.S)
  
dd.df <- melt(dd.df,  id.vars = 'cutoff', variable_name = 'series')

# plot on same grid, each series colored differently
ggplot(dd.df, aes(cutoff, value)) +
  geom_line(aes(colour = series)) +
  labs(x= "Number of OTU's", y="Slope of Distance-Decay") +
  labs(title="Dependence of Distance-Decay slope on rare taxa") +
  theme(legend.position="top") +
  guides(fill=guide_legend(title=NULL)) +
  theme(legend.title=element_blank())
```


## 4.) 

Little difference in community distance matrices between Active and All 
when using presence-absence, but generally high dissimilarity between all sites.
Clearer differences between Active and All based on Bray-Curtis, but also 
decreasing dissimilarity among sites. 

Notably, for 'all', a group of Hoosier National Forest (HNF) sites and a group of 
Brown County State Park (BC) sites share relatively high similarity. Also, the same 
subgroup of HNF sites is highly dissimilar to all other sites in the dataset. Perhaps 
these results are explained by environmental similarity.

```{r, results='hide', echo=F, message=F, warning=F, fig.width=7, fig.height=4}
#install.packages("gridExtra")
require(gridExtra) # also loads grid
require(lattice)

# Calculate Jaccard
active.Jaccard <- vegdist(active.comm, method = "bray", binary = TRUE)
all.Jaccard <- vegdist(all.comm, method = "bray", binary = TRUE)

# Calculate Bray-Curtis 
active.BrayCurtis <- vegdist(active.comm, method = "bray")
all.BrayCurtis <- vegdist(all.comm, method = "bray")

# Custome Color Palette
jet.colors <- colorRampPalette(c("#00007F", "blue", "#007FFF", "cyan", 
                                 "#7FFF7F", "yellow", "#FF7F00", "red",
                                 "#7F0000"))
# Define Order of Sites
order <- rev(attr(active.Jaccard, "Labels"))
grid <- expand.grid(x=order, y=order)

# Plot Heatmaps of Jaccard index
p1 <- levelplot(as.matrix(active.Jaccard)[, order], aspect = "iso", 
          col.regions = jet.colors, 
          xlab = "Pond", ylab = "Pond", 
          scales = list(cex = 0.5), 
          main = "Sorensen (active)")

p2 <- levelplot(as.matrix(all.Jaccard)[, order], aspect = "iso", 
          col.regions = jet.colors, 
          xlab = "Pond", ylab = "Pond", 
          scales = list(cex = 0.5), 
          main = "Sorensen (all)")

# Plot Heatmaps of Bray-Curtis distance
p3 <- levelplot(as.matrix(active.BrayCurtis)[, order], aspect = "iso",
          col.regions = jet.colors, 
          xlab = "Pond", ylab = "Pond", 
          scales = list(cex = 0.5), 
          main = "Bray-Curtis (active)")

p4 <- levelplot(as.matrix(all.BrayCurtis)[, order], aspect = "iso", 
          col.regions = jet.colors, 
          xlab = "Pond", ylab = "Pond", 
          scales = list(cex = 0.5), 
          main = "Bray-Curtis (all)")

grid.arrange(p1, p2, p3, p4, ncol=2, nrow=2)
```


## 5.) 
Cluster analysis reveals greater clustering based on relative activity 
than based on relative abundance.

```{r, results='hide', echo=F, message=F, warning=F, fig.width=7, fig.height=4}
# Set Initial Plot Parameters
plot.new()
par(mfrow=c(1, 1))#, mar = c(5, 4, 4, 2) + 0.1, oma =c(0,0,0,0))

# Perform Cluster Analysis
active.BC.ward <- hclust(active.BrayCurtis, method = "ward.D2")

# Plot Cluster
plot(active.BC.ward, main = "IN Ponds (Active)\nWard's Clustering", 
     ylab = "Squared Bray-Curtis Distance")
```

```{r, results='hide', echo=F, message=F, warning=F, fig.width=7, fig.height=4}
# Set Initial Plot Parameters
plot.new()
par(mfrow=c(1, 1))#, mar = c(5, 4, 4, 2) + 0.1, oma =c(0,0,0,0))

# Perform Cluster Analysis
all.BC.ward <- hclust(all.BrayCurtis, method = "ward.D2")

# Plot Cluster
plot(all.BC.ward, main = "IN Ponds (All)\nWard's Clustering", 
     ylab = "Squared Bray-Curtis Distance")
```


## 6.) 
Heat-mapping cluster analysis among dominant OTUs reveals that dominance 
in abundance does not reflect dominance in activity. Colormap represents 
log-transformed numbers of reads.

```{r, results='hide', echo=F, message=F, warning=F, fig.width=7, fig.height=8}
require(gplots)
plot.new()
all.comm.dom <- log10(all.comm[, 1:10]+1)
active.comm.dom <- log10(active.comm[, 1:10]+1)

# Set Initial Plot Parameters
par(mfrow=c(1, 1))#, mar = c(5, 4, 4, 2) + 0.1, oma =c(0,0,0,0))

heatmap.2(as.matrix(all.comm.dom), distfun = function(x) vegdist(x, method = "bray"),
          hclustfun = function(x) hclust(x, method = "ward.D2"), 
          col = jet.colors(100), trace = "none", density.info = "none",
          main='10 most dominant (All)')
```

```{r, results='hide', echo=F, message=F, warning=F, fig.width=7, fig.height=8}

# Set Initial Plot Parameters
plot.new()
par(mfrow=c(1, 1))
heatmap.2(as.matrix(active.comm.dom), 
          distfun = function(x) vegdist(x, method = "bray"),
          hclustfun = function(x) hclust(x, method = "ward.D2"), 
          col = jet.colors(100), trace = "none", 
          density.info = "none",
          main='10 most dominant (Active)')
```


## 7.) Variance partitioning: 
```{r, results='hide', echo=F, message=F, warning=F, fig.width=7, fig.height=4}

cutoffs <- c(15, 30, 60, 125, 250, 500, 1000, 2000, 4000, 8000, 16000, 22098)

active.env <- c()
active.geo <- c()
active.env.geo <- c()

all.env <- c()
all.geo <- c()
all.env.geo <- c()

# Geographic variables
geo.dat <- as.matrix(subset(env, select = lat:Elevation))
# Pond environmental variables
env.dat <- as.matrix(subset(env, select = Diameter:DON))

for (cutoff in cutoffs){
  print(cutoff)
  all.sub.comm <- all.comm[, 1:cutoff]
  active.sub.comm <- active.comm[, 1:cutoff]
  
  ## ALL
  #rda.geo <- rda(all.sub.comm ~ geo.dat)
  #rda.env <- rda(all.sub.comm ~ env.dat)
  
  # Two explanatory matrices -- Hellinger-transform Y
  mod <- varpart(Y=all.sub.comm, X=geo.dat, env.dat, transfo="hellinger")
  
  all.env <- c(all.env, mod[[1]][[3]][[3]][3])
  all.geo <- c(all.geo, mod[[1]][[3]][[3]][1])
  all.env.geo <- c(all.env.geo, mod[[1]][[3]][[3]][2])
  
  # Test fraction [a] using RDA:
  #geo.anova <- anova(rda.geo, step=200, perm.max=200)
  #geo.anova
  
  ## ACTIVE
  #rda.geo <- rda(active.sub.comm ~ geo.dat)
  #rda.env <- rda(active.sub.comm ~ env.dat)
  
  # Two explanatory matrices -- Hellinger-transform Y
  mod <- varpart(Y=active.sub.comm, X=geo.dat, env.dat, transfo="hellinger")
  
  active.env <- c(active.env, mod[[1]][[3]][[3]][3])
  active.geo <- c(active.geo, mod[[1]][[3]][[3]][1])
  active.env.geo <- c(active.env.geo, mod[[1]][[3]][[3]][2])
  }

# Set Initial Plot Parameters
plot.new()
par(mfrow=c(2, 2))

varpar.df <- data.frame(cutoff = cutoffs,
                 All_Env = all.env, 
                 All_Geography = all.geo,
                 All_GeoEnv = all.env.geo,
                 Active_Env = active.env, 
                 Active_Geography = active.geo,
                 Active_GeoEnv = active.env.geo)
  
varpar.df <- melt(varpar.df,  id.vars = 'cutoff', variable_name = 'series')

# plot on same grid, each series colored differently
ggplot(varpar.df, aes(cutoff, value)) +
  geom_line(aes(colour = series)) +
  labs(x= "Number of OTU's", y="Explained variance") +
  labs(title="Dependence of variance partitioning on rare taxa") +
  theme(legend.position="top") +
  guides(fill=guide_legend(title=NULL)) +
  theme(legend.title=element_blank())
```



### ii.  Constrained Ordination
Another way we can test hypotheses with continuous data is to use **constrained ordination**, which is sometimes referred to as canonical ordination.
Constrained ordination explores the relationships between two matrices: an **explanatory matrix** and a **response matrix**. 
Canonical correspondence analysis (CCA) and redundancy analysis (RDA) are two types of constrained ordination.
These techniques are based on the linear model framework and thus can be used to formally test hypotheses.
Constrained ordination works by first conducting multivariate multiple linear regression followed either by correspondence analysis (CA) with CCA or Principal Components Analysis (PCA) with RDA, while using the matrix of fitted values to obtain a constrained ordination.
A permutation test can then be used to test for overall significance. 
 
We will start by creating an explanatory matrix that contains water chemistry data.
We will then use the `cca()` function from the `vegan` package.
Note, we have to specify that we want the `cca` function in the `vegan` package because there are `cca` functions in both `vegan` and `ade4`!
We will then use permutation tests to evaluate the significance of our model. 
Finally, we will test the influence of each environmental variable on the constrained axes.

```{r, results = "hide"}
# Focus on dominant members
cutoff <- 50
all.sub.com <- all.comm[, 1:cutoff]
active.sub.com <- active.comm[, 1:cutoff]

# Define Environmental Matrix
env.chem <- env.dat

# Conduct CCA 
active.cca <- vegan::cca(active.sub.com ~ env.chem)

# Permutation Tests
anova(active.cca, by = "axis")
cca.fit <- envfit(active.cca, env.chem, perm = 999)
cca.fit

# Calculate Explained Variation
cca.explainvar1 <- round(active.cca$CCA$eig[1] / 
                         sum(c(active.cca$CCA$eig, active.cca$CA$eig)), 3) * 100
cca.explainvar2 <- round(active.cca$CCA$eig[2] / 
                         sum(c(active.cca$CCA$eig, active.cca$CA$eig)), 3) * 100

# Define Plot Parameters
par(mar = c(5, 5, 4, 4) + 0.1)

# Initiate Plot
plot(scores(active.cca, display = "wa"), xlim = c(-2.0, 1.5), ylim = c(-1.8, 2.5),
     xlab = paste("CCA 1 (", cca.explainvar1, "%)", sep = ""),
     ylab = paste("CCA 2 (", cca.explainvar2, "%)", sep = ""),
     pch = 16, cex = 1.0, type = "n", cex.lab = 1.5, cex.axis = 1.2, axes = FALSE)

# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
abline(h = 0, v = 0, lty = 3)
box(lwd = 2)

# Add Points & Labels
points(scores(active.cca, display = "wa"),
       pch = 19, cex = 3, bg = "gray", col = "gray")
text(scores(active.cca, display = "wa"), 
     labels = row.names(scores(active.cca, display = "wa")), cex=0.5)

# Add Environmental Vectors
vectors <- scores(all.cca, display = "bp")
row.names(vectors) <- colnames(env.dat)
arrows(0, 0, vectors[,1] * 2, vectors[, 2] * 2, 
       lwd = 1.5, lty = 1, length = 0.2, col = "red")
text(vectors[,1] * 2, vectors[, 2] * 2, pos = 3, 
     labels = row.names(vectors))
axis(side = 3, lwd.ticks=2, cex.axis=1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 1])) * 2, labels = pretty(range(vectors[, 1])))
axis(side = 4, lwd.ticks=2, cex.axis=1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 2])) * 2, labels = pretty(range(vectors[, 2])))
```


```{r, results = "hide"}
# Focus on dominant members
cutoff <- 50
all.sub.com <- all.comm[, 1:cutoff]

# Define Environmental Matrix
env.chem <- env.dat

# Conduct CCA 
all.cca <- vegan::cca(all.sub.com ~ env.chem)

# Permutation Tests
anova(all.cca, by = "axis")
cca.fit <- envfit(all.cca, env.chem, perm = 999)
cca.fit

# Calculate Explained Variation
cca.explainvar1 <- round(all.cca$CCA$eig[1] / 
                         sum(c(all.cca$CCA$eig, all.cca$CA$eig)), 3) * 100
cca.explainvar2 <- round(all.cca$CCA$eig[2] / 
                         sum(c(all.cca$CCA$eig, all.cca$CA$eig)), 3) * 100

# Define Plot Parameters
par(mar = c(5, 5, 4, 4) + 0.1)

# Initiate Plot
plot(scores(all.cca, display = "wa"), xlim = c(-2.0, 1.5), ylim = c(-2.2, 1.5),
     xlab = paste("CCA 1 (", cca.explainvar1, "%)", sep = ""),
     ylab = paste("CCA 2 (", cca.explainvar2, "%)", sep = ""),
     pch = 16, cex = 1.0, type = "n", cex.lab = 1.5, cex.axis = 1.2, axes = FALSE)#,
     #main = 'All (DNA)')

# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
abline(h = 0, v = 0, lty = 3)
box(lwd = 2)

# Add Points & Labels
points(scores(all.cca, display = "wa"),
       pch = 19, cex = 3, bg = "gray", col = "gray")
text(scores(all.cca, display = "wa"), 
     labels = row.names(scores(all.cca, display = "wa")), cex=0.5)

# Add Environmental Vectors
vectors <- scores(all.cca, display = "bp")
row.names(vectors) <- colnames(env.dat)
arrows(0, 0, vectors[,1] * 2, vectors[, 2] * 2, 
       lwd = 1.5, lty = 1, length = 0.2, col = "red")
text(vectors[,1] * 2, vectors[, 2] * 2, pos = 3, 
     labels = row.names(vectors))
axis(side = 3, lwd.ticks=2, cex.axis=1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 1])) * 2, labels = pretty(range(vectors[, 1])))
axis(side = 4, lwd.ticks=2, cex.axis=1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 2])) * 2, labels = pretty(range(vectors[, 2])))
```

