---
title: "Figure 1. Decay in compositional similarity with geographic and environmental distance"
geometry: margin=2.54cm
date: "November 7, 2015"
output: pdf_document
header-includes: \usepackage{array}
---

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Retrieve and set the working directory
rm(list=ls())
getwd()
setwd("~/GitHub/Dimensions/Aim3/papers/DD")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load packages
require("sp")          # Classes and methods for handling spatial data
require("geoR")        # Methods for geostatistical analyses
require("rgdal")       # Geospatial Data Abstraction Library
#require("raster")      # Methods to create a RasterLayer object
require("maptools")    # Tools for manipulating and reading geospatial data

#require("picante")
#require("ape")
#require("seqinr")
require("vegan") # biodiversity estimators and related functions
#require("fossil")
require("simba")
require("reshape")

require(reshape)

source("~/GitHub/Dimensions/Aim3/bin/MothurTools.R")
source("~/GitHub/Dimensions/Aim3/bin/metrics.R")
load(file = "~/GitHub/Dimensions/Aim3/Mothur/INPond_Initial.RData")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load Environmental and Geographical Data
env <- read.table("~/GitHub/Dimensions/Aim3/DATA/EnvData/20130801_PondDataMod.csv", sep = ",", header = TRUE)
lats <- as.numeric(env[, 3]) # latitudes (north and south)
lons <- as.numeric(env[, 4]) # longitudes (east and west)

# Geographic Distances (Kilometers) Among Ponds
long.lat <- as.matrix(cbind(env$long, env$lat))
coord.dist <- earth.dist(long.lat, dist = TRUE)
#coord.dist <- log(coord.dist)
coord.dist[which(!is.finite(coord.dist))] = NA
coord.dist.ls <- liste(coord.dist, entry = "geo.dist")

# Select DNA Data: Use the `grep()` Command and Rename with `gsub()`
# The active portion, based on cDNA 
active.comm <- Pond97[grep("*-cDNA", rownames(Pond97)), ]
rownames(active.comm) <- gsub("\\-cDNA", "", rownames(active.comm))
rownames(active.comm) <- gsub("\\_", "", rownames(active.comm))

# The community without respect to active or not, 16S rRNA gene sequences
all.comm <- Pond97[grep("*-DNA", rownames(Pond97)), ]
rownames(all.comm) <- gsub("\\-DNA", "", rownames(all.comm))
rownames(all.comm) <- gsub("\\_", "", rownames(all.comm))

# Remove Sites Not in the Environmental Data Set
active.comm <- active.comm[rownames(active.comm)  %in% env$Sample_ID, ]
all.comm <- all.comm[rownames(all.comm)  %in% env$Sample_ID, ]

# Remove Zero-Occurrence Taxa 
active.comm <- active.comm[ , colSums(active.comm) > 0]
all.comm <- all.comm[ , colSums(all.comm) > 0]

# N equals numbers of reads
# S equals the number of non-zero abundances
env$active.N <- as.vector(rowSums(active.comm)) 
env$active.S <- S.obs(active.comm)
env$active.H <- as.vector(diversity(active.comm, index="shannon"))
env$active.D <- as.vector(diversity(active.comm, index = "invsimpson")/env$active.S)

env$all.N <- as.vector(rowSums(all.comm)) 
env$all.S <- S.obs(all.comm)
env$all.H <- as.vector(diversity(all.comm, index="shannon"))
env$all.D <- as.vector(diversity(all.comm, index = "invsimpson")/env$all.S)
```


## Pattern 1: Distance-decay relationship

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE, fig.width=6, fig.height=6}
# Set Initial Plot Parameters

metric <- "jaccard"

# Taxonomic Distances Among Ponds (Bray-Curits)
active.weighted.dist <- 1 - vegdist(active.comm, method=metric, binary=FALSE) 
active.weighted.dist <- log(active.weighted.dist)
all.weighted.dist <- 1 - vegdist(all.comm, method=metric, binary=FALSE)
all.weighted.dist <- log(all.weighted.dist)

# Transform All Distances Into List Format:
active.weighted.dist.ls <- liste(active.weighted.dist, entry = metric)
all.weighted.dist.ls <- liste(all.weighted.dist, entry = metric)

# Create a Data Frame from the Lists of Distances
df <- data.frame(coord.dist.ls, active.weighted.dist.ls[, 3], all.weighted.dist.ls[, 3], active.presabs.dist.ls[, 3], all.presabs.dist.ls[, 3])

names(df)[4:7] <- c("active.weighted", "all.weighted", "active.env", "all.env")
attach(df)

#file <- paste("~/GitHub/Dimensions/Aim3/papers/DD/figs/", metric,"_microbes.png", sep="")
#png(filename=file)
  
#plot.new()
par(mfrow=c(2, 2), mar = c(5, 4, 3, 2) + 0.1, oma =c(0,1,3,0))

# Regression for active weighted
DD.active.weighted <- lm(active.weighted.dist ~ geo.dist)
summary(DD.active.weighted, correlation = TRUE)

# Regression for all weighted
DD.all.weighted <- lm(all.weighted.dist ~ geo.dist)
coeff <- summary(DD.all.weighted, correlation = TRUE)

# Regression for active env
DD.active.env <- lm(active.env.dist ~ geo.dist)
summary(DD.active.env, correlation = TRUE)

# Regression for all env
DD.all.presabs <- lm(all.presabs.dist ~ geo.dist)
summary(DD.all.presabs, correlation = TRUE)
  
  
# Make Plot for all weighted 
slope <- round(coefficients(DD.all.weighted)[2], 3)
p <- round(summary(DD.all.weighted)$coefficients[8], 3)
plot(coord.dist, all.weighted.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab="Distance, km", col = "azure3", cex.lab=1.5)
abline(DD.all.weighted , col = "red")
legend("bottomleft", paste("All (weighted)\nslope=",slope," p = ",p), bty="n", cex=1.2)

# Make Plot for active weighted
slope <- round(coefficients(DD.active.weighted)[2], 3)
p <- round(summary(DD.active.weighted)$coefficients[8],3)
plot(coord.dist, active.weighted.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab="Distance, km",col = "azure3",cex.lab=1.5)
abline(DD.active.weighted , col = "red")
legend("bottomleft", paste("Active (weighted)\nslope=",slope," p = ",p), bty="n", cex=1.2)

# Make Plot for all env
slope <- round(coefficients(DD.all.presabs)[2], 3)
p <- round(summary(DD.all.presabs)$coefficients[8],3)
plot(coord.dist, all.presabs.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab="Distance, km", col = "azure3",cex.lab=1.5)
abline(DD.all.presabs , col = "red")
legend("bottomleft", paste("All (unweighted)\nslope=",slope," p = ",p), bty="n", cex=1.2)

# Make Plot for active env
slope <- round(coefficients(DD.active.presabs)[2], 3)
p <- round(summary(DD.active.presabs)$coefficients[8],3)
plot(coord.dist, active.presabs.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab="Distance, km", col = "azure3",cex.lab=1.5)
abline(DD.active.presabs , col = "red")
legend("bottomleft", paste("Active (unweighted)\nslope = ",slope," p = ",p), bty="n", cex=1.2)

# Add X-Axis Label to Plot
mtext("Geographic Distance, km", side = 1, adj = 0, outer = TRUE)
  
d1 <- diffslope(geo.dist, active.weighted, geo.dist, all.weighted)
d2 <- diffslope(geo.dist, active.presabs, geo.dist, all.presabs)

  
if(metric == "morisita"){
  metric <- "Morisita"
  }else if(metric == "mountford"){
    metric <- "Mountford"
  }else if(metric == "jaccard"){
    metric <- "Jaccard"
  }else if(metric == "chao"){
    metric <- "Chao"
  }else if(metric == "canberra"){
    metric <- "Canberra"
  }else if(metric == "bray"){
    metric <- "Bray"
  }else if(metric == "kulczynski"){
    metric <- "Kulczynski"
  }else if(metric == "horn"){
    metric <- "Horn"
  }
  
  Mtitle <- paste(metric,"\n",'Weighted: Difference in slope =', 
                  round(as.numeric(d1[2]),3), '; p =', d1[3],"\n",
                  'Unweighted: Difference in slope =', 
                  round(as.numeric(d2[2]),3), '; p =', d2[3])

  title(main=Mtitle, line=-2, outer=T, cex.main=1.5)
  
  #dev.off()
}
```

```{r, results = 'hide', message = FALSE, warning = FALSE}

struc.dist <- 1 - vegdist(OTUs) # Bray-Curtis similarity between the plots
coord.dist <- dist(as.matrix(lats, lons)) # geographical distance between plots

# transform environmental data to numeric types
temp <- as.numeric(Ponds$"Salinity")
elev <- as.numeric(Ponds$"ORP")
depth <- as.numeric(Ponds$"Depth")
doc <- as.numeric(Ponds$"DOC")

# calculate the distance (Euclidean) between the plots regarding environmental variables
env.dist <- 1 - vegdist(cbind(temp, elev, depth, doc), "euclidean")

# transform all distance matrices into list format:
struc.dist.ls <- liste(struc.dist, entry="struc")
env.dist.ls <- liste(env.dist, entry="env")
coord.dist.ls <- liste(coord.dist, entry="dist")
```

Now, create a data frame containing similarity of the environment and similarity of community.

```{r}
df <- data.frame(coord.dist.ls, env.dist.ls[,3], struc.dist.ls[,3])
names(df)[4:5] <- c("env", "struc")
attach(df) #df <- subset(df, struc != 0)
```

Finally, let's plot the Distance-decay relationships, with regression lines in red.

```{r}
par(mfrow=c(1, 1))
plot(dist, env, xlab="Geographic Distance", ylab="Euclidean Distance", 
     main = "Distance-Decay for the Environment", col='SteelBlue')

OLS <- lm(env ~ dist)
OLS # print regression results to the screen
abline(OLS, col="red4")

par(mfrow=c(1, 1))
plot(dist, struc, xlab="Geographic Distance", ylab="Bray-Curtis (similarity)", 
     main="Distance-Decay for Community Composition", col='darkorchid4')

OLS <- lm(struc ~ dist)
OLS # print regression results to the screen
abline(OLS, col="red4")
```

Let's, examine the slope of the regression lines, asking whether they are significantly different from one another.

```{r}
diffslope(dist, env, dist, struc)
```

***Question 3***: Are microbial communities that are closer in geographic distance also closer in compositional similarity? How about for environmental condition?

> ***Answer 3***:



## Concept 2: Spatial Aggregation
Tobler made a general observation that occurs in nearly all systems, i.e., spatial autocorrelation.
A related observation is that natural phenomena are generally clustered, i.e., spatially aggregated.
That is, individuals, conditions, and events often occur in patches, clusters, pulses, etc.
Take for example, the ponds in our sample area.
A highly level of aggregation would suggest that if we encounter one individual, then we are likely to encounter others of the same species nearby.

## Pattern 2: Spatial abundance distribution
One of the primary patterns of spatial aggregation in ecology is the distribution of a species abundance within a landscape, also referred to as the **species spatial abundance distribution (SSAD)**.
The SSAD reveals the frequency at which we find a species at a particular abundance.
In this way, the SSAD is similar to a histogram.

Here, we will examine SSADs for OTU's in the refuge pond dataset by constructing **kernel-density curves**.  
Kernel density curves are analogous to histograms, but avoid the arbitrary creation of bins or discrete classes.
In constructing kernel density curves, we attempt to account for uncertainty and sampling error by focusing on the probability that a randomly drawn data point will take a value within a particular range, instead of the exact frequencies we observed.

For example, suppose we were interested in how the location of individuals varied across sites or samples.
Let's simulate this by drawing values from a normal distribution, at random.

```{r}
S.list <- rnorm(1000) # 1000 randomly drawn values from a normal distribution
plot(density(S.list), col = 'red', xlab='Location',
     ylab='Probability Density',  main = 'A kernel density curve for abundance across space')
```


Below, we will examine the SSADs of OTUs that are randomly drawn from the refuge ponds dataset. 
But first, let's begin by defining a function that will generate the SSAD for a randomly drawn OTU.

```{r}
ssad <- function(x){
  ad <- c(2, 2)
  ad <- OTUs[, otu]
  ad = as.vector(t(x = ad))
  ad = ad[ad > 0]
}
```

Next, we will draw 4 OTUs at random and plot their SSADs.
But first, we will need to introduce **while loops**.

### While loops
If you have ever heard anything to the effect of, "While you're at it, do this...", then you are familiar with the concept of a while loop.
The while loop is a type of **control flow** structure that allows us to tell a program to perform an operation *while* some condition is satisfied.

For example, we might want R to draw numbers at random until 4 numbers less than 50 have been drawn.
```{r}
numbers = c() 
while (length(numbers) < 4){ # while the counter is less than 4
  x <- runif(1, 1, 100) # draw a number at random from 1 to 100
  if (x < 50){    # if the number is less than 50...
    numbers <- c(numbers, x)
  }
}
numbers # check our numbers, each should be less than 50
```


Having very briefly intodroduced while loops, and inaverdently, an `if` statement, lets write a chunk of code that will draw OTUs at random from our refuge ponds dataset, and then generate their spatial abundance distributions (i.e. SSADs).

```{r}
par(mfrow=c(2, 2))

ct <- 0         # a counter variable
while (ct < 4){ # the while statement
  otu <- sample(1:length(OTUs), 1) # choose a random OTU
  ad <- ssad(otu)                  # find the OTU's SSAD
  if (length(ad) > 10 & sum(ad > 100)){ 
    ct <- ct + 1
    plot(density(ad), col = 'magenta', xlab='Site abundance',
    ylab='Probability Density',  main = otu.names[otu])
    }
  }
```

Feel free to run this chunk as many time as you like.

***Question 4***: Is the sampled abundance for a given OTU often aggregated? If so, how do you know, that is, how do you interpret the pattern in the kernel density curve? Are there many sites with low abundance and few sites with high abundance?

> ***Answer 4***:

***Question ***: The SSAD is a statistical distribution and like all statistical distributions, the SSAD has a mean, median, mode, variance, skewness, kurtosis, etc. What aspects of the SSAD would you choose to quantify and report, and why? 


***Question 5***: Each row in the site-by-species matrix represents a site. 
Each column represents an OTU.
If the SSAD is generated by considering all rows for a single column (i.e. OTU), then what do we obtain when we consider all columns for a given row (i.e. site)?
Have we examined this sort of data structure before? 

> ***Answer 5***: 


## Concept 3: Scale-Dependence

Our idea of whether variables are spatially autocorrelated and whether the abundances of OTUs are spatially aggregated can change with aspects of spatial scale, i.e. extent and grain.
**Extent** is the greatest distance considered in an observation or study.
**Grain** is the smallest or primary unit by which the extent is measured.

Let's generate two random samples from a normal distribution, one sample for x-coordinates and one for y-coordinates.
We'll let each x-y pair represent the location of a single individual, where all individuals belong to the same species.
Then, we'll plot the spatial distribution of our randomly distributed individuals at different extents.

```{r}
par(mfrow=c(2, 2))

x <- rnorm(20000)
y <- rnorm(20000)

plot(x,y, xlim=c(-4, 4), ylim=c(-4, 4), pch=".", col='Steelblue')
plot(x,y, xlim=c(-2, 2), ylim=c(-2, 2), pch=".", col='Steelblue')
plot(x,y, xlim=c(-1, 1), ylim=c(-1, 1), pch=".", col='Steelblue')
plot(x,y, xlim=c(-0.5, 0.5), ylim=c(-0.5, 0.5), pch=".", col='Steelblue')
```

***Question 6***: What effect does changing the extent have on aggregation? Do you find this important or interesting given that 1.) all points were drawn from the same distribution and 2.) each plot contains the same points as all other plots with smaller extent?

> ***Answer 6***:

It should be clear from above, that 'random' does not mean absent of aggregation.
In fact, most statistical distributions from which random samples can be drawn are very aggregated.
That is, they have obvious modes around which most values occur.

Moving on, let's explore the effect of changing spatial `grain`, from a fine grain to a coarse grain. 
We will do this while holding extent constant and will plot heat maps (i.e. 2D histogram) revealing the density of individuals in the landscape.
We will then plot kernel density curves to reveal the probability that an individual chosen at random from the landscape will have come from a site with a particular abundance.

```{r, results = 'hide', message = FALSE, warning = FALSE}
require("gplots")
par(mfrow=c(2, 2))

df <- data.frame(x,y)

h1 <- hist2d(df, nbins=80, show=TRUE, xlim=c(-2,2), ylim=c(-2,2),
             xlab='x-coord', ylab='y-coord', main = "Fine grain" )
ad <- h1$counts
plot(density(ad), col = 'red', xlab='Site abundance',
     ylab='Probability Density',  main = "Fine grain")

h4 <- hist2d(df, nbins=20, show=TRUE, xlim=c(-2,2), ylim=c(-2,2),
             xlab='x-coord', ylab='y-coord', main = "Coarse grain" )
ad <- h4$counts
plot(density(ad), col = 'red', xlab='Site abundance',
     ylab='Probability Density',  main = "Coarse grain")
```

***Question 7***: Beyond changing the pixilated appearance of the plots, what does changing the spatial grain mean for interpreting aggregation? Consider the kernel density plots. 

> ***Answer 7***:



***Question 8***: How are the kernel density curves we just generated for our randomly drawn points related to the species spatial abundance distributions (SSAD) that we generated for OTUs in our refuge plots?

> ***Answer 8***:



## Primary Concept 3: Spatial Accumulation

So far, we have discussed spatial autocorrelation and aggregation as core concepts of geographical ecology.
Likewise, we have introduced and examined primary patterns for both of those concepts.
Here, we introduce another core concept, accumulation across space.
It may seem self-evident that, if starting from the scale of a single individual and increasing our sample area, that we will inevitably encounter more species, OTUs, or other taxa.

For example, suppose we replicate our above random sampling strategy of drawing x-y coordinates from a normal distribution.
But, instead of drawing just one sample representing one species, we will draw 50 samples, with each representing a species with 1000 individuals.

```{r, results = 'hide', message = FALSE, warning = FALSE}
community <- c()
species <- c()

# initiate the plot
plot(0, 0, col='white', xlim = c(0, 100), ylim = c(0, 100),
     xlab='x coordinate', ylab='y coordinate', 
     main='A simulated landscape occupied by 100
     species, having 1000 individuals apiece.')

while (length(community) < 100){ # while our community has less than 100 species
  # choosing the mean, standard deviation, and colors at random
  std <- runif(1, 1, 10)
  x <- rnorm(1000, mean = runif(1, 0, 100), sd = std) 
  y <- rnorm(1000, mean = runif(1, 0, 100), sd = std)
  color <- c(rgb(runif(1),runif(1),runif(1)))
  
  points(x, y, pch=".", col=color)
  species <- list(x, y, color)
  community[[length(community)+1]] <- species
  }
```

Having generated a simulated landscape occupied by 50 species having 1000 individuals apiece, we can examine how richness can accumulate with area. 
Let's begin by picking a corner at random and then accumulating area.

```{r}
lim <- 10

S.list <- c()
A.list <- c()

while (lim <= 100){ # while the spatial extent on the x and y is less than or equal to 100
  S <- 0            # Set richness to be zero
  for (sp in community){ # for each species in the community
    xs <- sp[[1]] # assign the x coordinates
    ys <- sp[[2]] # assing the y coordinates
    sp.name <- sp[[3]]  # assign the species name
    xy.coords <- cbind(xs, ys) 
    for (xy in xy.coords){ # for each pair of xy coordinates in xy.coords
      if (max(xy) <= lim){ # if the individual is within our current spatial extent...
        S <- S + 1         # then the species occurs there
          break  # break out of the last for loop because we're only considering 
                 # incidence, and not abundance. 
                 # In other words, if the species occurs once, that's good enough
        }
      }
    }
  S.list <- c(S.list, log10(S))
  A.list <- c(A.list, log10(lim^2))
  lim <- lim * 1.5  # increase the extent multiplicately, but slowly
}
```


Having generated our primary vectors, S.list and A.list, we can analyze how species richness scales with area. In short, we can analyze one of ecology's oldest and most intensively studied patterns, the **Species-Area Relationship**.


```{r, results = 'hide', message = FALSE, warning = FALSE}
results <- lm(S.list ~ A.list)
plot(A.list, S.list, col="dark red", pch=20, cex=2, 
     main="Species-area relationship",
     xlab='ln(Area)', ylab='ln(Richness)')

abline(results, col="red", lwd=2)

int <- round(results[[1]][[1]],2)
z <- round(results[[1]][[2]],2)
legend(x=2.1, y=2, c('slope = ', z))
```


## Pattern 3: Species-area relationship (SAR)

The fact that we accumulate species, and likewise increase richness, with increasing area is far from interesting.
In fact, we just showed that we can expect this as a result of random sampling.
What is interesting is the rate at which the accumulation of taxa occurs.
Arrhenius (1921) first described the general form of the *species-area relationship (SAR)* as a power-law: $S = cA^{z}$ where S is species richnness and A is area.

Power-laws reveal how one quantity scales with another, most often across orders of magnitude.
Arrhenius's formula predicts a rate of increase in richness that is approximately linear in log-log space. That is, $log(S) = c + zlog(A)$, where z is the scaling exponent.

***Question 9***: The authors of your assigned reading revealed that the exponent of the SAR may be influenced by geographic, ecological, and evolutionary factors. But, what in general, is the value of z?

> ***Answer 9***:


***Question 10***: What was the slope of the species-area relationship for our randomly assembled community? Is this similar to the slopes you encountered in the reading?

> ***Answer 10***:


***Question 11***: We could use this 'random placement' approach to model how many ecological phenomena might occur via random sampling. What other spatial aspects of alpha and beta diversity could we address? Suggest at least 3.

> ***Answer 11***:



## 7) HOMEWORK

**1.)** Complete the in-class exercise, Knit to a pdf, and submit a pull request.

**2.)** Each refuge pond has an associated diameter. Build the species-area relationship for the refuge pond dataset using the following recipe: 

1. Using the formula for the area of a circle, calculate the area for each pond. 
2. Randomly choose one pond and obtain its area and richness.
3. Choose two ponds at random and obtain their combined richness and their summed area.
Do not simply sum the richnesses of the two sites, as this will result in double-counting species.
4. Choose three, four, five, etc. ponds at random, repeating the above steps each time.
You will eventually work you way up to 51 ponds. **You will need to use loops for this**.
5. At this point you should have two vectors, one for richness (S) and one for area (A).
6. Plot the SAR and estimate its slope.
7. REPEAT steps 2 through 6 one thousand times, adding each new SAR to the same plot.
Once again, you will need to use loops, as above.
8. In a second plot, generate a kernel density curve for the slopes of the SARs.

**3.)** Draw several general conclusions from your analyses in question #2.

**4.)** Which environmental and diversity variables reveal positive spatial autocorrelation?

**5.)** A. How many OTUs are present at more than 10 sites? How many OTUs only occur at one site?

**6.)** In considering total abundances (N) among the refuge ponds, we are really only considering the number of detected 16S rRNA reads for any given OTU.
Find the mode of the SSAD for each OTU that is present at 10 or more sites; it's difficult to generate an informative histogram for less than 10 sites.
Then, generate a kernel density curve for these modes, revealing the pattern of modal abundance across these more common OTUs. 
Draw general conclusions about trends across the refuge pond OTUs from your results.
