---
title: "Supplemental Figures: Compositional similarity vs. Environmental distance"
geometry: margin=2.54cm
date: "November 7, 2015"
output: pdf_document
header-includes: \usepackage{array}
---

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Retrieve and set the working directory
rm(list=ls())
getwd()
setwd("~/GitHub/Dimensions/Aim3/papers/DD")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load packages
require("sp")          # Classes and methods for handling spatial data
require("geoR")        # Methods for geostatistical analyses
require("rgdal")       # Geospatial Data Abstraction Library
require("raster")      # Methods to create a RasterLayer object
require("maptools")    # Tools for manipulating and reading geospatial data

require("picante")
require("ape")
require("seqinr")
require("vegan") # biodiversity estimators and related functions
require("fossil")
require("simba")
require("reshape")
require(ggplot2)
require(reshape)

source("~/GitHub/Dimensions/Aim3/bin/MothurTools.R")
source("~/GitHub/Dimensions/Aim3/bin/metrics.R")
load(file = "~/GitHub/Dimensions/Aim3/Mothur/INPond_Initial.RData")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE, fig.width=6, fig.height=6}
# Load Environmental and Geographical Data
env.org <- read.table("~/GitHub/Dimensions/Aim3/DATA/EnvData/20130801_PondDataMod.csv",
                      sep = ",", header = TRUE)

env.org <- subset(env.org, chla < 2000)
env.org <- subset(env.org, pH > 1)
env.org <- subset(env.org, Salinity > 0.0)
env.org <- subset(env.org, TDS < 5.0)

env.names <- subset(env.org, select=c(Depth:DON))
#env.names <- subset(env.org, select=c(TDS))
env.names <- names(env.names)

for(name in env.names){

  metric <- "jaccard"
  #x.lab <- paste("log(difference in ",name,")", sep="")
  x.lab <- paste("Difference in ",name)
  
  env <- subset(env.org, select=c(name))
  # calculate the distance (Euclidean) between the plots regarding environmental variables
  env.dist <- vegdist(env, "euclidean")
  #env.dist[env.dist == 0] <- 0.001
  #env.dist <- log10(env.dist)
  
  env.dist.ls <- liste(env.dist, entry="env.dist")

  # Select DNA Data: Use the `grep()` Command and Rename with `gsub()`
  # The active portion, based on cDNA 
  active.comm <- Pond97[grep("*-cDNA", rownames(Pond97)), ]
  rownames(active.comm) <- gsub("\\-cDNA", "", rownames(active.comm))
  rownames(active.comm) <- gsub("\\_", "", rownames(active.comm))

  # The community without respect to active or not, 16S rRNA gene sequences
  all.comm <- Pond97[grep("*-DNA", rownames(Pond97)), ]
  rownames(all.comm) <- gsub("\\-DNA", "", rownames(all.comm))
  rownames(all.comm) <- gsub("\\_", "", rownames(all.comm))

  # Remove Sites Not in the Environmental Data Set
  active.comm <- active.comm[rownames(active.comm)  %in% env.org$Sample_ID, ]
  all.comm <- all.comm[rownames(all.comm)  %in% env.org$Sample_ID, ]

  write.table(active.comm, 
              file = "~/GitHub/Dimensions/Aim3/DATA/ForPython/ActiveComm", 
              sep = ",", eol = "\n", na = "NA", dec = ".", 
              row.names = TRUE, col.names = TRUE)
  
  write.table(all.comm, 
              file = "~/GitHub/Dimensions/Aim3/DATA/ForPython/AllComm", 
              sep = ",", eol = "\n", na = "NA", dec = ".", 
              row.names = TRUE, col.names = TRUE)
  
  # Remove Zero-Occurrence Taxa 
  active.comm <- active.comm[ , colSums(active.comm) > 0]
  all.comm <- all.comm[ , colSums(all.comm) > 0]

  file <- paste("~/GitHub/Dimensions/Aim3/papers/DD/figs/", metric,"_ComDist-",name,".png", sep="")
  
  # Taxonomic Distances Among Ponds (Bray-Curits)
  active.weighted.dist <- 1 - vegdist(active.comm, method=metric, binary=FALSE) 
  active.weighted.dist <- log10(active.weighted.dist)
  all.weighted.dist <- 1 - vegdist(all.comm, method=metric, binary=FALSE)
  all.weighted.dist <- log10(all.weighted.dist)

  active.presabs.dist <- 1 - vegdist(active.comm, method=metric, binary=TRUE) 
  active.presabs.dist <- log10(active.presabs.dist)
  all.presabs.dist <- 1 - vegdist(all.comm, method=metric, binary=TRUE)
  all.presabs.dist <- log10(all.presabs.dist)

  # Transform All Distances Into List Format:
  active.weighted.dist.ls <- liste(active.weighted.dist, entry = metric)
  all.weighted.dist.ls <- liste(all.weighted.dist, entry = metric)
  active.presabs.dist.ls <- liste(active.presabs.dist, entry = metric)
  all.presabs.dist.ls <- liste(all.presabs.dist, entry = metric)

  # Create a Data Frame from the Lists of Distances
  df <- data.frame(env.dist.ls, active.weighted.dist.ls[, 3], all.weighted.dist.ls[, 3], active.presabs.dist.ls[, 3], all.presabs.dist.ls[, 3])

  names(df)[4:7] <- c("active.weighted", "all.weighted", "active.presabs", "all.presabs")
  attach(df)

  png(filename=file)  
  plot.new()
  par(mfrow=c(2, 2), mar = c(5, 4, 3, 2) + 0.1, oma =c(0,1,3,0))

  # Regression for active weighted
  DD.active.weighted <- lm(active.weighted ~ env.dist)
  summary(DD.active.weighted, correlation = TRUE)

  # Regression for all weighted
  DD.all.weighted <- lm(all.weighted ~ env.dist)
  coeff <- summary(DD.all.weighted, correlation = TRUE)

  # Regression for active presence/absence
  DD.active.presabs <- lm(active.presabs ~ env.dist)
  summary(DD.active.presabs, correlation = TRUE)

  # Regression for all presence/absence 
  DD.all.presabs <- lm(all.presabs ~ env.dist)
  summary(DD.all.presabs, correlation = TRUE)
  
  # Make Plot for all weighted 
  slope <- round(coefficients(DD.all.weighted)[2], 3)
  p <- round(summary(DD.all.weighted)$coefficients[8], 3)
  plot(env.dist, all.weighted.dist, xaxt = "s", las = 1, 
     ylab="log(similarity)", xlab=x.lab, col = "azure3", cex.lab=1.5)
  abline(DD.all.weighted , col = "red")
  legend("bottomleft", paste("All (weighted)\nslope=",slope," p = ",p), bty="n", cex=1.2)

  # Make Plot for active weighted
  slope <- round(coefficients(DD.active.weighted)[2], 3)
  p <- round(summary(DD.active.weighted)$coefficients[8],3)
  plot(env.dist, active.weighted.dist, xaxt = "s", las = 1, 
     ylab="log(similarity)", xlab=x.lab,col = "azure3",cex.lab=1.5)
  abline(DD.active.weighted , col = "red")
  legend("bottomleft", paste("Active (weighted)\nslope=",slope," p = ",p), bty="n", cex=1.2)

  # Make Plot for all presence/absence
  slope <- round(coefficients(DD.all.presabs)[2], 3)
  p <- round(summary(DD.all.presabs)$coefficients[8],3)
  plot(env.dist, all.presabs.dist, xaxt = "s", las = 1, 
     ylab="log(similarity)", xlab=x.lab, col = "azure3",cex.lab=1.5)
  abline(DD.all.presabs , col = "red")
  legend("bottomleft", paste("All (unweighted)\nslope=",slope," p = ",p), bty="n", cex=1.2)

  # Make Plot for active presence/absence
  slope <- round(coefficients(DD.active.presabs)[2], 3)
  p <- round(summary(DD.active.presabs)$coefficients[8],3)
  plot(env.dist, active.presabs.dist, xaxt = "s", las = 1, 
     ylab="log(similarity)", xlab=x.lab, col = "azure3",cex.lab=1.5)
  abline(DD.active.presabs , col = "red")
  legend("bottomleft", paste("Active (unweighted)\nslope = ",slope," p = ",p), bty="n", cex=1.2)

  # Add X-Axis Label to Plot
  mtext("Geographic Distance, km", side = 1, adj = 0, outer = TRUE)
  
  d1 <- diffslope(env.dist, active.weighted, env.dist, all.weighted)
  d2 <- diffslope(env.dist, active.presabs, env.dist, all.presabs)

  metric <- "Jaccard"
  
  Mtitle <- paste(metric,"\n",'Weighted: Difference in slope =', 
                  round(as.numeric(d1[2]),3), '; p =', d1[3],"\n",
                  'Unweighted: Difference in slope =', 
                  round(as.numeric(d2[2]),3), '; p =', d2[3])

  title(main=Mtitle, line=-2, outer=T, cex.main=1.5)
  
  dev.off()
}
```
