---
title: "Spatial vs. temporal storage: The dual role of microbial seed
banks in driving geographical patterns of microbial diversity"

geometry: margin=2.54cm
date: "August 4, 2015"
output: pdf_document
header-includes: \usepackage{array}
---

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Retrieve and set the working directory
rm(list=ls())
getwd()
setwd("~/GitHub/Dimensions/Aim3")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load packages
require("sp")          # Classes and methods for handling spatial data
require("geoR")        # Methods for geostatistical analyses
require("rgdal")       # Geospatial Data Abstraction Library
require("raster")      # Methods to create a RasterLayer object
require("maptools")    # Tools for manipulating and reading geospatial data

require("picante")
require("ape")
require("seqinr")
require("vegan") # biodiversity estimators and related functions
require("fossil")
require("simba")
require("reshape")

require(ggplot2)
require(reshape)
```



```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Import functions
# import code file that includes a function for reading in output files 
#from community sequencing software `mothur` (http://www.mothur.org/).
source("~/GitHub/Dimensions/Aim3/bin/MothurTools.R")
load(file = "~/GitHub/Dimensions/Aim3/Mothur/INPond_Initial.RData")
```



```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load Environmental and Geographical Data
env <- read.table("~/GitHub/Dimensions/Aim3/DATA/EnvData/20130801_PondDataMod.csv",
                  sep = ",", header = TRUE)

# Select DNA Data: Use the `grep()` Command and Rename with `gsub()`
# The active portion, based on cDNA 
active.comm <- Pond97[grep("*-cDNA", rownames(Pond97)), ]
rownames(active.comm) <- gsub("\\-cDNA", "", rownames(active.comm))
rownames(active.comm) <- gsub("\\_", "", rownames(active.comm))

# The community without respect to active or not, 16S rRNA gene sequences
all.comm <- Pond97[grep("*-DNA", rownames(Pond97)), ]
rownames(all.comm) <- gsub("\\-DNA", "", rownames(all.comm))
rownames(all.comm) <- gsub("\\_", "", rownames(all.comm))

# Remove Sites Not in the Environmental Data Set
active.comm <- active.comm[rownames(active.comm)  %in% env$Sample_ID, ]
all.comm <- all.comm[rownames(all.comm)  %in% env$Sample_ID, ]

# Remove Zero-Occurrence Taxa 
active.comm <- active.comm[ , colSums(active.comm) > 0]
all.comm <- all.comm[ , colSums(all.comm) > 0]

# Geographic variables
geo.dat <- as.matrix(subset(env, select = lat:Elevation))
# Pond environmental variables
env.dat <- as.matrix(subset(env, select = Diameter:DON))
```

### Constrained Ordination
Constrained ordination explores the relationships between two matrices: an **explanatory matrix** and a **response matrix**. 
Canonical correspondence analysis (CCA) and redundancy analysis (RDA) are two types of constrained ordination.
Constrained ordination works by first conducting multivariate multiple linear regression followed either by correspondence analysis (CA) with CCA or Principal Components Analysis (PCA) with RDA, while using the matrix of fitted values to obtain a constrained ordination.
A permutation test can then be used to test for overall significance. 

We will start by creating an explanatory matrix that contains water chemistry data.
We will then use the `cca()` function from the `vegan` package.
Note, we have to specify that we want the `cca` function in the `vegan` package because there are `cca` functions in both `vegan` and `ade4`!
We will then use permutation tests to evaluate the significance of our model. 
Finally, we will test the influence of each environmental variable on the constrained axes.

```{r, results = "hide"}
# Get Dominant members
cutoff <- 1000

all.sub.comm <- all.comm[, 1:cutoff]
active.sub.comm <- active.comm[, 1:cutoff]

#all.sub.comm <- all.comm
#active.sub.comm <- active.comm

# Define Environmental Matrix
env.chem <- env.dat

# Conduct CCA 
active.cca <- vegan::cca(active.sub.comm ~ env.chem)

# Permutation Tests
anova(active.cca, by = "axis")
cca.fit <- envfit(active.cca, env.chem, perm = 999)
cca.fit

# Calculate Explained Variation
cca.explainvar1 <- round(active.cca$CCA$eig[1] / 
                         sum(c(active.cca$CCA$eig, active.cca$CA$eig)), 3) * 100
cca.explainvar2 <- round(active.cca$CCA$eig[2] / 
                         sum(c(active.cca$CCA$eig, active.cca$CA$eig)), 3) * 100

# Define Plot Parameters
par(mar = c(5, 5, 4, 4) + 0.1)

# Initiate Plot
plot(scores(active.cca, display = "wa"), xlim = c(-3.5, 2), ylim = c(-3.2, 3.2),
     xlab = paste("CCA 1 (", cca.explainvar1, "%)", sep = ""),
     ylab = paste("CCA 2 (", cca.explainvar2, "%)", sep = ""),
     pch = 16, cex = 2.0, type = "n", cex.lab = 1.5, cex.axis = 1.2, axes = FALSE)

# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
abline(h = 0, v = 0, lty = 3)
box(lwd = 2)

# Add Points & Labels
points(scores(active.cca, display = "wa"),
       pch = 19, cex = 3, bg = "gray", col = "gray")
text(scores(active.cca, display = "wa"), 
     labels = row.names(scores(active.cca, display = "wa")))

# Add Environmental Vectors
vectors <- scores(active.cca, display = "bp")
row.names(vectors) <- c("diameter", "depth", "volume", "ORP", "temp", "SpC", "DO", "TDS", "salinity", "pH", "color", "chla", "DOC", "DON")

arrows(0, 0, vectors[,1] * 2, vectors[, 2] * 2, 
       lwd = 2, lty = 1, length = 0.2, col = "red")
text(vectors[,1] * 2, vectors[, 2] * 2, pos = 3, 
     labels = row.names(vectors))
axis(side = 3, lwd.ticks=2, cex.axis=1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 1])) * 2, labels = pretty(range(vectors[, 1])))
axis(side = 4, lwd.ticks=2, cex.axis=1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 2])) * 2, labels = pretty(range(vectors[, 2])))
```


```{r, results = "hide"}
cca.scores <- scores(active.cca)
cca.sites <- as.data.frame(scores(cca.scores$sites))
cca1.scores <- as.matrix(subset(cca.sites, select = CCA1))

cca1.dist <- vegdist(cca1.scores, "euclidean")
cca1.dist.ls <- liste(cca1.dist, entry="cca1.dist")
```



```{r, results = "hide"}
metric <- "jaccard"
x.lab <- "Difference in CCA1 score"


file <- paste("~/GitHub/Dimensions/Aim3/papers/DD/figs/", metric,"_CCA1_Dist.png", sep="")
  
# Taxonomic Distances Among Ponds (Bray-Curits)
active.weighted.dist <- 1 - vegdist(active.comm, method=metric, binary=FALSE) 
active.weighted.dist <- log10(active.weighted.dist)
all.weighted.dist <- 1 - vegdist(all.comm, method=metric, binary=FALSE)
all.weighted.dist <- log10(all.weighted.dist)

active.presabs.dist <- 1 - vegdist(active.comm, method=metric, binary=TRUE) 
active.presabs.dist <- log10(active.presabs.dist)
all.presabs.dist <- 1 - vegdist(all.comm, method=metric, binary=TRUE)
all.presabs.dist <- log10(all.presabs.dist)

# Transform All Distances Into List Format:
active.weighted.dist.ls <- liste(active.weighted.dist, entry = metric)
all.weighted.dist.ls <- liste(all.weighted.dist, entry = metric)
active.presabs.dist.ls <- liste(active.presabs.dist, entry = metric)
all.presabs.dist.ls <- liste(all.presabs.dist, entry = metric)

# Create a Data Frame from the Lists of Distances
df <- data.frame(cca1.dist.ls, active.weighted.dist.ls[, 3], all.weighted.dist.ls[, 3], active.presabs.dist.ls[, 3], all.presabs.dist.ls[, 3])

names(df)[4:7] <- c("active.weighted", "all.weighted", "active.presabs", "all.presabs")
attach(df)

png(filename=file)  
plot.new()
par(mfrow=c(2, 2), mar = c(5, 4, 3, 2) + 0.1, oma =c(0,1,3,0))

# Regression for active weighted
DD.active.weighted <- lm(active.weighted ~ cca1.dist)
summary(DD.active.weighted, correlation = TRUE)

# Regression for all weighted
DD.all.weighted <- lm(all.weighted ~ cca1.dist)
coeff <- summary(DD.all.weighted, correlation = TRUE)

# Regression for active presence/absence
DD.active.presabs <- lm(active.presabs ~ cca1.dist)
summary(DD.active.presabs, correlation = TRUE)

# Regression for all presence/absence 
DD.all.presabs <- lm(all.presabs ~ cca1.dist)
summary(DD.all.presabs, correlation = TRUE)
  
# Make Plot for all weighted 
slope <- round(coefficients(DD.all.weighted)[2], 3)
p <- round(summary(DD.all.weighted)$coefficients[8], 3)
plot(cca1.dist, all.weighted.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab=x.lab, col = "azure3", cex.lab=1.5)
abline(DD.all.weighted , col = "red")
legend("topright", paste("All (weighted)\nslope=",slope," p = ",p), bty="n", cex=1.2)

# Make Plot for active weighted
slope <- round(coefficients(DD.active.weighted)[2], 3)
p <- round(summary(DD.active.weighted)$coefficients[8],3)
plot(cca1.dist, active.weighted.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab=x.lab,col = "azure3",cex.lab=1.5)
abline(DD.active.weighted , col = "red")
legend("topright", paste("Active (weighted)\nslope=",slope," p = ",p), bty="n", cex=1.2)
  
# Make Plot for all presence/absence
slope <- round(coefficients(DD.all.presabs)[2], 3)
p <- round(summary(DD.all.presabs)$coefficients[8],3)
plot(cca1.dist, all.presabs.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab=x.lab, col = "azure3",cex.lab=1.5)
abline(DD.all.presabs , col = "red")
legend("bottomleft", paste("All (unweighted)\nslope=",slope," p = ",p), bty="n", cex=1.2)

# Make Plot for active presence/absence
slope <- round(coefficients(DD.active.presabs)[2], 3)
p <- round(summary(DD.active.presabs)$coefficients[8],3)
plot(cca1.dist, active.presabs.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab=x.lab, col = "azure3",cex.lab=1.5)
abline(DD.active.presabs , col = "red")
legend("bottomleft", paste("Active (unweighted)\nslope = ",slope," p = ",p), bty="n", cex=1.2)

# Add X-Axis Label to Plot
mtext("Geographic Distance, km", side = 1, adj = 0, outer = TRUE)
  
d1 <- diffslope(cca1.dist, active.weighted, cca1.dist, all.weighted)
d2 <- diffslope(cca1.dist, active.presabs, cca1.dist, all.presabs)

metric <- "Jaccard"
  
Mtitle <- paste(metric,"\n",'Weighted: Difference in slope =', 
                round(as.numeric(d1[2]),3), '; p =', d1[3],"\n",
                'Unweighted: Difference in slope =', 
                round(as.numeric(d2[2]),3), '; p =', d2[3])

title(main=Mtitle, line=-2, outer=T, cex.main=1.5)
  
dev.off()
```



