---
title: "Spatial vs. temporal storage: The dual role of microbial seed
banks in driving geographical patterns of microbial diversity"

geometry: margin=2.54cm
date: "August 4, 2015"
output: pdf_document
header-includes: \usepackage{array}
---

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Retrieve and set the working directory
rm(list=ls())
getwd()
setwd("~/GitHub/Dimensions/Aim3")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load packages
require("picante")
require("ape")
require("seqinr")
require("vegan") # biodiversity estimators and related functions
require("fossil")
require("simba")
require("reshape")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Import functions
# import code file that includes a function for reading in output files 
#from community sequencing software `mothur` (http://www.mothur.org/).
source("~/GitHub/Dimensions/Aim3/bin/MothurTools.R")
load(file = "~/GitHub/Dimensions/Aim3/Mothur/INPond_Initial.RData")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load Environmental and Geographical Data
env <- read.table("~/GitHub/Dimensions/Aim3/DATA/EnvData/20130801_PondDataMod.csv",
                  sep = ",", header = TRUE)

env <- subset(env, chla < 2000)
env <- subset(env, pH > 1)
env <- subset(env, Salinity > 0.0)
env <- subset(env, TDS < 5.0)

# Geographic Distances (Kilometers) Among Ponds
long.lat <- as.matrix(cbind(env$long, env$lat))
geo.dist <- earth.dist(long.lat, dist = TRUE)
#geo.dist <- log(geo.dist)
geo.dist[which(!is.finite(geo.dist))] = NA
geo.dist.ls <- liste(geo.dist, entry = "geo.dist")


# Select DNA Data: Use the `grep()` Command and Rename with `gsub()`
# The active portion, based on cDNA 
active.comm <- Pond97[grep("*-cDNA", rownames(Pond97)), ]
rownames(active.comm) <- gsub("\\-cDNA", "", rownames(active.comm))
rownames(active.comm) <- gsub("\\_", "", rownames(active.comm))

# The community without respect to active or not, 16S rRNA gene sequences
all.comm <- Pond97[grep("*-DNA", rownames(Pond97)), ]
rownames(all.comm) <- gsub("\\-DNA", "", rownames(all.comm))
rownames(all.comm) <- gsub("\\_", "", rownames(all.comm))

# Remove Sites Not in the Environmental Data Set
active.comm <- active.comm[rownames(active.comm)  %in% env$Sample_ID, ]
all.comm <- all.comm[rownames(all.comm)  %in% env$Sample_ID, ]

# Remove Zero-Occurrence Taxa 
active.comm <- active.comm[ , colSums(active.comm) > 0]
all.comm <- all.comm[ , colSums(all.comm) > 0]

#cutoff <- 4000
#all.comm <- all.comm[, 1:cutoff]
#active.comm <- active.comm[, 1:cutoff]

# Geographic variables
geo.dat <- as.matrix(subset(env, select = lat:long))
# Pond environmental variables
env.dat <- as.matrix(subset(env, select = Diameter:DON))
```



```{r}
levels.short <- c('BCSP', 'HNF', 'YSF')
colors.short <- c('firebrick1', 'darkorchid1', 'cornflowerblue')
  
colors <- c('firebrick1','firebrick1','firebrick1','firebrick1','firebrick1',
            'firebrick1','firebrick1','firebrick1','firebrick1','firebrick1',
            'firebrick1','firebrick1','firebrick1','firebrick1','firebrick1',
            'firebrick1',
            'darkorchid1','darkorchid1','darkorchid1','darkorchid1','darkorchid1',
            'darkorchid1','darkorchid1','darkorchid1','darkorchid1','darkorchid1',
            'darkorchid1','darkorchid1','darkorchid1','darkorchid1','darkorchid1',
            'darkorchid1',
            'cornflowerblue','cornflowerblue','cornflowerblue','cornflowerblue',
            'cornflowerblue', 'cornflowerblue','cornflowerblue','cornflowerblue',
            'cornflowerblue','cornflowerblue', 'cornflowerblue','cornflowerblue',
            'cornflowerblue','cornflowerblue','cornflowerblue')

env.dat <- as.matrix(subset(env, select = Diameter:DON))
env.chem <- env.dat
# Conduct CCA 
active.cca <- vegan::cca(active.comm ~ env.chem)
all.cca <- vegan::cca(all.comm ~ env.chem)
```



```{r}
file <- paste("~/GitHub/Dimensions/Aim3/papers/DD/DD/figs/CCA-active.png", sep="")
png(filename=file)

cca.explainvar1 <- round(active.cca$CCA$eig[1] / 
                         sum(c(active.cca$CCA$eig, active.cca$CA$eig)), 3) * 100
cca.explainvar2 <- round(active.cca$CCA$eig[2] / 
                         sum(c(active.cca$CCA$eig, active.cca$CA$eig)), 3) * 100

# Define Plot Parameters
par(mfrow=c(1, 1), mar = c(5, 5, 4, 4) + 0.1, pty="s")

# Initiate Plot
plot(scores(active.cca, display = "wa"), xlim = c(-2, 2.7), ylim = c(-2.5, 2.5),
     xlab = paste("CCA 1 (", cca.explainvar1, "%)", sep = ""),
     ylab = paste("CCA 2 (", cca.explainvar2, "%)", sep = ""),
     pch = 16, cex = 2.0, type = "n", cex.lab = 1.5, cex.axis = 1.2,
     axes = FALSE)

# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
abline(h = 0, v = 0, lty = 3)
box(lwd = 2)

# Add Points & Labels
points(scores(active.cca, display = "wa"),
       pch = 1, cex = 2, bg = colors, col = colors)
legend("topleft", legend = levels.short, bty = "n", col = colors.short, 
       pch = 1)

# Add Environmental Vectors
vectors <- scores(active.cca, display = "bp")
row.names(vectors) <- c("diameter", "depth", "volume", "ORP", "temp",
      "SpC", "DO", "TDS", "salinity", "pH", "color", "chla", "DOC", "DON")

arrows(0, 0, vectors[,1] * 3, vectors[, 2] * 3, 
       lwd = 1, lty = 1, length = 0.2, col = "black")

text(vectors[,1] * 3.5, vectors[, 2] * 3.5, pos = 3, 
     labels = row.names(vectors))

axis(side = 3, lwd.ticks=2, cex.axis=1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 1])) * 2, labels = pretty(range(vectors[, 1])))
axis(side = 4, lwd.ticks=2, cex.axis=1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 2])) * 2, labels = pretty(range(vectors[, 2])))

dev.off()
```



```{r}
file <- paste("~/GitHub/Dimensions/Aim3/papers/DD/DD/figs/CCA-all.png", sep="")
png(filename=file)

# Permutation Tests
#anova(active.cca, by = "axis")
#cca.fit <- envfit(active.cca, env.chem, perm = 999)

# Calculate Explained Variation
cca.explainvar1 <- round(all.cca$CCA$eig[1] / 
                         sum(c(all.cca$CCA$eig, all.cca$CA$eig)), 3) * 100
cca.explainvar2 <- round(all.cca$CCA$eig[2] / 
                         sum(c(all.cca$CCA$eig, all.cca$CA$eig)), 3) * 100

# Define Plot Parameters
par(mfrow=c(1, 1), mar = c(5, 5, 4, 4) + 0.1, pty="s")

# Initiate Plot
plot(scores(all.cca, display = "wa"), xlim = c(-1.5, 3), ylim = c(-1.5, 2.0),
     xlab = paste("CCA 1 (", cca.explainvar1, "%)", sep = ""),
     ylab = paste("CCA 2 (", cca.explainvar2, "%)", sep = ""),
     pch = 16, cex = 2.0, type = "n", cex.lab = 1.5, 
     cex.axis = 1.2, axes = FALSE)

# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
abline(h = 0, v = 0, lty = 3)
box(lwd = 2)

# Add Points & Labels
points(scores(all.cca, display = "wa"),
       pch = 1, cex = 2, bg = colors, col = colors)
legend("topleft", legend = levels.short, bty = "n", 
       col = colors.short, pch = 1)

# Add Environmental Vectors
vectors <- scores(all.cca, display = "bp")
row.names(vectors) <- c("diameter", "depth", "volume", "ORP", 
                        "temp", "SpC", "DO", "TDS", "salinity",
                        "pH", "color", "chla", "DOC", "DON")

arrows(0, 0, vectors[,1] * 3, vectors[, 2] * 3, 
       lwd = 1, lty = 1, length = 0.2, col = "black")

text(vectors[, 1] * 3.5, vectors[, 2] * 3.5, pos = 3, 
     labels = row.names(vectors))

axis(side = 3, lwd.ticks=2, cex.axis=1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 1])) * 2, labels = pretty(range(vectors[, 1])))
axis(side = 4, lwd.ticks=2, cex.axis=1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 2])) * 2, labels = pretty(range(vectors[, 2])))

dev.off()
```



```{r, results = "hide"}
active.cca.scores <- scores(active.cca)
active.cca.sites <- as.data.frame(scores(active.cca.scores$sites))
active.cca.scores <- as.matrix(subset(active.cca.sites, select = CCA1))

active.cca.dist <- vegdist(active.cca.scores, "euclidean")
#active.cca.dist.ls <- liste(cca1.dist, entry="active.cca1.dist")

all.cca.scores <- scores(all.cca)
all.cca.sites <- as.data.frame(scores(all.cca.scores$sites))
all.cca.scores <- as.matrix(subset(all.cca.sites, select = CCA1))

all.cca.dist <- vegdist(all.cca.scores, "euclidean")
#all.cca1.dist.ls <- liste(cca1.dist, entry="all.cca1.dist")
```



```{r, results = "hide"}
metric <- "bray"
x.lab <- "Difference in CCA1 score"

file <- paste("~/GitHub/Dimensions/Aim3/papers/DD/DD/figs/", 
              metric,"_CCA1_Dist.png", sep="")
  
png(filename=file)  
plot.new()
par(mfrow=c(2, 2), mar = c(5, 5, 4, 2) + 0.1, oma=c(0,0,2,0))

# Taxonomic Distances Among Ponds (Bray-Curits)
active.weighted.dist <- 1 - vegdist(active.comm, method=metric, binary=FALSE) 
#active.weighted.dist <- log10(active.weighted.dist)
all.weighted.dist <- 1 - vegdist(all.comm, method=metric, binary=FALSE)
#all.weighted.dist <- log10(all.weighted.dist)

active.presabs.dist <- 1 - vegdist(active.comm, method=metric, binary=TRUE) 
#active.presabs.dist <- log10(active.presabs.dist)
all.presabs.dist <- 1 - vegdist(all.comm, method=metric, binary=TRUE)
#all.presabs.dist <- log10(all.presabs.dist)


# Regression for active weighted
DD.active.weighted <- lm(active.weighted.dist ~ active.cca.dist)

# Regression for all weighted
DD.all.weighted <- lm(all.weighted.dist ~ all.cca.dist)

# Regression for active presence/absence
DD.active.presabs <- lm(active.presabs.dist ~ active.cca.dist)

# Regression for all presence/absence 
DD.all.presabs <- lm(all.presabs.dist ~ all.cca.dist)
  
# Make Plot for all weighted 
slope <- round(coefficients(DD.all.weighted)[2], 3)
p <- round(summary(DD.all.weighted)$coefficients[8], 3)
plot(all.cca.dist, all.weighted.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab=x.lab, col = "azure3", cex.lab=1.5)
abline(DD.all.weighted , col = "red")
legend("topright", paste("All (weighted)\nslope=",slope," p = ",p), 
       bty="n", cex=1.2)

# Make Plot for active weighted
slope <- round(coefficients(DD.active.weighted)[2], 3)
p <- round(summary(DD.active.weighted)$coefficients[8],3)
plot(active.cca.dist, active.weighted.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab=x.lab,col = "azure3",cex.lab=1.5)
abline(DD.active.weighted , col = "red")
legend("topright", paste("Active (weighted)\nslope=",slope," p = ",p),
       bty="n", cex=1.2)
  
# Make Plot for all presence/absence
slope <- round(coefficients(DD.all.presabs)[2], 3)
p <- round(summary(DD.all.presabs)$coefficients[8],3)
plot(all.cca.dist, all.presabs.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab=x.lab, col = "azure3",cex.lab=1.5)
abline(DD.all.presabs , col = "red")
legend("bottomleft", paste("All (unweighted)\nslope=",slope," p = ",p),
       bty="n", cex=1.2)

# Make Plot for active presence/absence
slope <- round(coefficients(DD.active.presabs)[2], 3)
p <- round(summary(DD.active.presabs)$coefficients[8],3)
plot(active.cca.dist, active.presabs.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab=x.lab, col = "azure3",cex.lab=1.5)
abline(DD.active.presabs , col = "red")
legend("bottomleft", paste("Active (unweighted)\nslope = ", 
                          slope," p = ",p), bty="n", cex=1.2)

# Add X-Axis Label to Plot
mtext("Geographic Distance, km", side = 1, adj = 0, outer = TRUE)
  
d1 <- diffslope(active.cca.dist, active.weighted, all.cca.dist, all.weighted)
d2 <- diffslope(active.cca.dist, active.presabs, all.cca.dist, all.presabs)

Mtitle <- paste('Weighted: Difference in slope =', 
                round(as.numeric(d1[2]),3), '; p =', d1[3],"\n",
                'Unweighted: Difference in slope =', 
                round(as.numeric(d2[2]),3), '; p =', d2[3])

title(main=Mtitle, line=-2, outer=T, cex.main=1.5)
dev.off()
```



```{r, results = "hide"}
metric <- "bray"
x.lab <- "Geographic distance, km"

file <- paste("~/GitHub/Dimensions/Aim3/papers/DD/DD/figs/",
              metric,"_Geo_Dist.png", sep="")
  
png(filename=file)  
plot.new()
par(mfrow=c(2, 2), mar = c(5, 5, 4, 2) + 0.1, oma=c(0,0,2,0))

# Taxonomic Distances Among Ponds (Bray-Curits)
active.weighted.dist <- 1 - vegdist(active.comm, method=metric, binary=FALSE) 
#active.weighted.dist <- log10(active.weighted.dist)
all.weighted.dist <- 1 - vegdist(all.comm, method=metric, binary=FALSE)
#all.weighted.dist <- log10(all.weighted.dist)

active.presabs.dist <- 1 - vegdist(active.comm, method=metric, binary=TRUE) 
#active.presabs.dist <- log10(active.presabs.dist)
all.presabs.dist <- 1 - vegdist(all.comm, method=metric, binary=TRUE)
#all.presabs.dist <- log10(all.presabs.dist)

# Regression for active weighted
DD.active.weighted <- lm(active.weighted.dist ~ geo.dist)

# Regression for all weighted
DD.all.weighted <- lm(all.weighted.dist ~ geo.dist)

# Regression for active presence/absence
DD.active.presabs <- lm(active.presabs.dist ~ geo.dist)

# Regression for all presence/absence 
DD.all.presabs <- lm(all.presabs.dist ~ geo.dist)
  
# Make Plot for all weighted 
slope <- round(coefficients(DD.all.weighted)[2], 3)
p <- round(summary(DD.all.weighted)$coefficients[8], 3)
plot(geo.dist, all.weighted.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab=x.lab, col = "azure3", 
   cex.lab=1.5, ylim=c(0, 1))
abline(DD.all.weighted , col = "red")
legend("topright", paste("All (weighted)\nslope=",slope," p = ",p), 
       bty="n", cex=1.2)

# Make Plot for active weighted
slope <- round(coefficients(DD.active.weighted)[2], 3)
p <- round(summary(DD.active.weighted)$coefficients[8],3)
plot(geo.dist, active.weighted.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab=x.lab,col = "azure3",
   cex.lab=1.5, ylim=c(0, 1))
abline(DD.active.weighted , col = "red")
legend("topright", paste("Active (weighted)\nslope=",slope," p = ",p),
       bty="n", cex=1.2)
  
# Make Plot for all presence/absence
slope <- round(coefficients(DD.all.presabs)[2], 3)
p <- round(summary(DD.all.presabs)$coefficients[8],3)
plot(geo.dist, all.presabs.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab=x.lab, col = "azure3",
   cex.lab=1.5, ylim=c(0, 0.5))
abline(DD.all.presabs , col = "red")
legend("bottomleft", paste("All (unweighted)\nslope=",slope," p = ",p),
       bty="n", cex=1.2)

# Make Plot for active presence/absence
slope <- round(coefficients(DD.active.presabs)[2], 3)
p <- round(summary(DD.active.presabs)$coefficients[8],3)
plot(geo.dist, active.presabs.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab=x.lab, col = "azure3",
   cex.lab=1.5, ylim=c(0, 0.5))
abline(DD.active.presabs , col = "red")
legend("bottomleft", paste("Active (unweighted)\nslope = ", 
                          slope," p = ",p), bty="n", cex=1.2)

# Add X-Axis Label to Plot
mtext("Geographic Distance, km", side = 1, adj = 0, outer = TRUE)
  
d1 <- diffslope(active.cca.dist, active.weighted, active.cca.dist, all.weighted)
d2 <- diffslope(active.cca.dist, active.presabs, active.cca.dist, all.presabs)

Mtitle <- paste('Weighted: Difference in slope =', 
                round(as.numeric(d1[2]),3), '; p =', d1[3],"\n",
                'Unweighted: Difference in slope =', 
                round(as.numeric(d2[2]),3), '; p =', d2[3])

title(main=Mtitle, line=-2, outer=T, cex.main=1.5)
dev.off()
```