---
title: "Spatial vs. temporal storage: The dual role of microbial seed
banks in driving geographical patterns of microbial diversity"

geometry: margin=2.54cm
date: "August 4, 2015"
output: pdf_document
header-includes: \usepackage{array}
---

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Retrieve and set the working directory
rm(list=ls())
getwd()
setwd("~/GitHub/Dimensions/Aim3")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load packages
require("picante")
require("ape")
require("seqinr")
require("vegan") # biodiversity estimators and related functions
require("fossil")
require("simba")
require("reshape")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Import functions
# import code file that includes a function for reading in output files 
#from community sequencing software `mothur` (http://www.mothur.org/).
source("~/GitHub/Dimensions/Aim3/bin/MothurTools.R")
load(file = "~/GitHub/Dimensions/Aim3/Mothur/INPond_Initial.RData")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load Environmental and Geographical Data
env <- read.table("~/GitHub/Dimensions/Aim3/DATA/EnvData/20130801_PondDataMod.csv",
                  sep = ",", header = TRUE)

# Remove 189, 190, 191
# 236, 237 missing lots of data (very big ponds, outliers of sorts)

# HNF276 Meg fixed pH data


#env <- subset(env, chla < 2000)
env <- subset(env, pH > 1)
#env <- subset(env, Salinity > 0.0)
env <- subset(env, TDS < 5.0)

# Geographic Distances (Kilometers) Among Ponds
long.lat <- as.matrix(cbind(env$long, env$lat))
geo.dist <- earth.dist(long.lat, dist = TRUE)
geo.dist[which(!is.finite(geo.dist))] = NA

# Select DNA Data: Use the `grep()` Command and Rename with `gsub()`
# The active portion, based on cDNA 
active.comm <- Pond97[grep("*-cDNA", rownames(Pond97)), ]
rownames(active.comm) <- gsub("\\-cDNA", "", rownames(active.comm))
rownames(active.comm) <- gsub("\\_", "", rownames(active.comm))

# The community without respect to active or not, 16S rRNA gene sequences
all.comm <- Pond97[grep("*-DNA", rownames(Pond97)), ]
rownames(all.comm) <- gsub("\\-DNA", "", rownames(all.comm))
rownames(all.comm) <- gsub("\\_", "", rownames(all.comm))

# Remove Sites Not in the Environmental Data Set
active.comm <- active.comm[rownames(active.comm)  %in% env$Sample_ID, ]
all.comm <- all.comm[rownames(all.comm)  %in% env$Sample_ID, ]

# Remove Zero-Occurrence Taxa 
active.comm <- active.comm[ , colSums(active.comm) > 0]
all.comm <- all.comm[ , colSums(all.comm) > 0]

cutoff <- 1000
all.comm <- all.comm[, 1:cutoff]
active.comm <- active.comm[, 1:cutoff]

# Geographic variables
geo.dat <- as.matrix(subset(env, select = lat:long))
# Pond environmental variables
env.dat <- as.matrix(subset(env, select = Diameter:DON))
```


```{r}
levels.short <- c('BCSP', 'HNF', 'YSF')
colors.short <- c('firebrick1', 'darkorchid1', 'cornflowerblue')
  
colors <- c('firebrick1','firebrick1','firebrick1','firebrick1','firebrick1',
            'firebrick1','firebrick1','firebrick1','firebrick1','firebrick1',
            'firebrick1','firebrick1','firebrick1','firebrick1','firebrick1',
            'firebrick1',
            'darkorchid1','darkorchid1','darkorchid1','darkorchid1','darkorchid1',
            'darkorchid1','darkorchid1','darkorchid1','darkorchid1','darkorchid1',
            'darkorchid1','darkorchid1','darkorchid1','darkorchid1','darkorchid1',
            'darkorchid1',
            'cornflowerblue','cornflowerblue','cornflowerblue','cornflowerblue',
            'cornflowerblue', 'cornflowerblue','cornflowerblue','cornflowerblue',
            'cornflowerblue','cornflowerblue', 'cornflowerblue','cornflowerblue',
            'cornflowerblue','cornflowerblue','cornflowerblue')

env.dat <- as.matrix(subset(env, select = Diameter:DON))
env.chem <- env.dat

# Standardize and center
env.chem <- scale(env.chem, center = TRUE, scale = TRUE)

# Conduct PCA 
pca <- princomp(env.chem)
scores <- as.data.frame(pca$scores[,1:2])
```



```{r}
#file <- paste("~/GitHub/Dimensions/Aim3/papers/DD/DD/figs/CCA-active.png", sep="")
#png(filename=file)

# Define Plot Parameters
#par(mfrow=c(1, 1), mar = c(5, 5, 4, 4) + 0.1, pty="s")

# Initiate Plot
#dev.off()
```



```{r, results = "hide"}
#pca.sites <- as.data.frame(scores(pca.scores$sites))
#pca.scores <- as.matrix(subset(pca.sites, select = CCA1))
pca.dist <- vegdist(scores, "euclidean")
```



```{r, results = "hide"}
metric <- "bray"
x.lab <- "Difference in PCA1 score"

file <- paste("~/GitHub/Dimensions/Aim3/papers/DD/DD/figs/", 
              metric,"_PCA1_Dist.png", sep="")
  
png(filename=file)  
plot.new()
par(mfrow=c(1, 2), mar = c(5, 5, 2, 2) + 0.1, oma=c(0,0,2,0), pty="s")

active.presabs.dist <- 1 - vegdist(active.comm, method=metric, binary=TRUE) 
all.presabs.dist <- 1 - vegdist(all.comm, method=metric, binary=TRUE)

# Regression for active presence/absence
DD.active.presabs <- lm(active.presabs.dist ~ pca.dist)

# Regression for all presence/absence 
DD.all.presabs <- lm(all.presabs.dist ~ pca.dist)
  
# Make Plot for all presence/absence
slope <- round(coefficients(DD.all.presabs)[2], 3)
p <- round(summary(DD.all.presabs)$coefficients[8],3)
plot(pca.dist, all.presabs.dist, xaxt = "s", las = 1, 
   ylab="Community similarity", xlab=x.lab, col = "azure4",
   cex.lab=1)
abline(DD.all.presabs , col = "red", lwd = 3)
legend("topleft", paste("All\nslope=",slope),
       bty="n", cex=1)

# Make Plot for active presence/absence
slope <- round(coefficients(DD.active.presabs)[2], 3)
p <- round(summary(DD.active.presabs)$coefficients[8],3)
plot(pca.dist, active.presabs.dist, xaxt = "s", las = 1, 
   ylab="Community similarity", xlab=x.lab, col = "azure4",
   cex.lab=1)
abline(DD.active.presabs , col = "red", lwd = 3)
legend("topleft", paste("Active\nslope = ", slope),
       bty="n", cex=1)

# Add X-Axis Label to Plot
mtext("Geographic Distance, km", side = 1, adj = 0, outer = TRUE)
  
#d <- diffslope(pca.dist, active.presabs.dist, pca.dist, all.presabs.dist)


#Mtitle <- paste('Unweighted: Difference in slope =', 
#                round(as.numeric(d2[2]),3), '; p =', d2[3])
#print(Mtitle)
#title(main=Mtitle, line=-2, outer=T, cex.main=1.5)
dev.off()
```



```{r, results = "hide"}
metric <- "bray"
x.lab <- "Geographic distance, km"

file <- paste("~/GitHub/Dimensions/Aim3/papers/DD/DD/figs/",
              metric,"_Geo_Dist.png", sep="")
  
png(filename=file)  
plot.new()
par(mfrow=c(1, 2), mar = c(5, 5, 2, 2) + 0.1, oma=c(0,0,2,0), pty="s")


# Taxonomic Distances Among Ponds
active.presabs.dist <- 1 - vegdist(active.comm, method=metric, binary=TRUE) 
all.presabs.dist <- 1 - vegdist(all.comm, method=metric, binary=TRUE)

# Regression for active presence/absence
DD.active.presabs <- lm(active.presabs.dist ~ geo.dist)

# Regression for all presence/absence 
DD.all.presabs <- lm(all.presabs.dist ~ geo.dist)
  
# Make Plot for all presence/absence
slope <- round(coefficients(DD.all.presabs)[2], 3)
p <- round(summary(DD.all.presabs)$coefficients[8],3)
plot(geo.dist, all.presabs.dist, xaxt = "s", las = 1, 
   ylab="Community similarity", xlab=x.lab, col = "azure4",
   cex.lab=1, ylim=c(0.1, 0.6))
abline(DD.all.presabs , col = "red", lwd = 3)
legend("topleft", paste("All\nslope=",slope),
       bty="n", cex=1)

# Make Plot for active presence/absence
slope <- round(coefficients(DD.active.presabs)[2], 3)
p <- round(summary(DD.active.presabs)$coefficients[8],3)
plot(geo.dist, active.presabs.dist, xaxt = "s", las = 1, 
   ylab="Community similarity", xlab=x.lab, col = "azure4",
   cex.lab=1, ylim=c(0.1, 0.6))
abline(DD.active.presabs , col = "red", lwd = 3)
legend("topleft", paste("Active\nslope = ", 
                          slope), bty="n", cex=1)

# Add X-Axis Label to Plot
mtext("Geographic Distance, km", side = 1, adj = 0, outer = TRUE)
  
#d1 <- diffslope(active.cca.dist, active.weighted, active.cca.dist, all.weighted)
d2 <- diffslope(active.cca.dist, active.presabs.dist, active.cca.dist, all.presabs.dist)

dev.off()
```