---
title: "Dimensions of Biodiversity - Aim 1, Persistence"
author: "Jay T. Lennon and Stuart E. JOnes"
date: "`r format(Sys.time(), '%d %B, %Y')`"
header-includes:
   - \usepackage{array}
output: pdf_document
geometry: margin=2.54cm
---

## 1) SETUP
### A. Retrieve and Set Your Working Directory

```{r, results = 'hide'}
rm(list = ls())
getwd()
setwd("~/GitHub/Dimensions/Aim1/DeathCurves/Phylo")
```

### B. Load Packages 

```{r, results = 'hide', message = FALSE, warning = FALSE} 
require("muscle")
require("seqinr")
require("ape")

require("phylobase")
require("adephylo")
require("geiger") 
require("picante")
require("stats")
require("RColorBrewer")
require("caper")
```

## 2) Use mothur alingment based on Silva reference to make a tree
### A. Performing mothur alignment

i) Copy FASTA file from AFS to Mason
Open terminal and type the following commands: 
    ```sh
    ssh karst.uits.iu.edu
    cd /afs/iu.edu/home/l/e/lennon/Lennon_Shared/Long-term_Dormancy/Sequences
    kinit
    aklog
    cd /afs/iu.edu/home/l/e/lennon/Lennon_Shared/Long-term_Dormancy/Sequences
    cp persistence.fasta /N/dc2/projects/Lennon_Sequences/Persistence
    ```

ii) Perform mothur alingnment
    ```sh
    ssh lennonj@mason.indiana.edu
    cd /N/dc2/projects/Lennon_Sequences/Persistence
    module load gcc
    module load mothur/1.31.2
    mothur
    align.seqs(fasta=persistence.fasta, reference=silva.bacteria.fasta, flip=T, processors=4)
    summary.seqs(fasta=persistence.align)
    screen.seqs(fasta=persistence.align, minlength=758) 
    filter.seqs(fasta=persistence.good.align, vertical=T, trump=.)
    quit()
    ```
Open new terminal and cd to Github project folder.  
Move the file using the following commands:
    ```sh
    #scp lennonj@karst.uits.iu.edu:/N/dc2/projects/Lennon_Sequences/Persistence/persistence.align ./
    #scp lennonj@karst.uits.iu.edu:/N/dc2/projects/Lennon_Sequences/Persistence/persistence.align.report ./
    scp lennonj@karst.uits.iu.edu:/N/dc2/projects/Lennon_Sequences/Persistence/persistence.good.filter.fasta ./
    ```
### B. Visualize mothur alignment
```{r}
# Read Alignment File {seqinr}
read.aln.M <- read.alignment(file = "./persistence.good.filter.fasta", format = "fasta")  

# Convert Alignment File to DNAbin Object {ape}
p.DNAbin.M <- as.DNAbin(read.aln.M) 

# Identify Base Pair Region of 16S rRNA Gene to Visuzlize
window.M <- p.DNAbin.M[, 0:321] 

# Command to Visusalize Sequence Alignment {ape}
image.DNAbin(window.M, cex.lab = 0.50) 
```

## 3) Make mothur-aligned tree
# Read Alignment File {seqinr}
```{r}
# Create Distance Matrix with "raw" Model {ape}
seq.dist.raw.M <- dist.dna(p.DNAbin.M, model = "raw", pairwise.deletion = FALSE)

# Neighbor Joining Algorithm to Construct Tree, a 'phylo' Object {ape}
nj.tree.M <- bionjs(seq.dist.raw.M)

# Identify Outgroup Sequence
outgroup.M <- match("Methanosarcina", nj.tree.M$tip.label)

# Root the Tree {ape}
nj.rooted.M <- root(nj.tree.M, outgroup.M, resolve.root = TRUE)

# Load phylo taxonomy data
tax <- read.table("persistence.phylo.txt", sep = "\t", header = TRUE)
tax.name<-paste(tax$Code,tax$Genus)
nj.tree.M$tip.label <- match(nj.tree.M$tip.label,tax.name)

# Plot the Rooted Tree{ape}
par(mar = c(1,1,2,1) + 0.1)
plot.phylo(nj.rooted.M, main = "Neigbor Joining Tree from mothur Alignment", 
           "phylogram", use.edge.length = FALSE, direction = "right",
           cex = 0.6, label.offset = 1, show.tip.label = FALSE, x.lim = 30)
           
tiplabels(tax.name, adj = c(0,0.5), cex = 0.5, frame = "none",
          pch = NULL, thermo = NULL, pie = NULL, piecol = NULL,
          col = NULL, bg = NULL)

add.scale.bar(cex = 0.7)
```