---
title: "The geographical ecology of pond bacteria"
geometry: margin=2.54cm
date: "April 21, 2015"
output: pdf_document
header-includes: \usepackage{array}
---

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Retrieve and set the working directory
rm(list=ls())
getwd()
setwd("~/GitHub/Dimensions/Aim3")
```

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load Packages 
require("sp")          # Classes and methods for handling spatial data
require("geoR")        # Methods for geostatistical analyses
require("rgdal")       # Geospatial Data Abstraction Library
require("raster")      # Methods to create a RasterLayer object
require("maptools")    # Tools for manipulating and reading geospatial data

require("picante")
require("ape")
require("seqinr")
require("vegan") # biodiversity estimators and related functions
require("fossil")
require("simba")
require("reshape")

# We used a source code file that includes a function for reading in the output files from the popular community sequencing software `mothur` (http://www.mothur.org/).
source("~/GitHub/Dimensions/Aim3/bin/MothurTools.R")

load(file = "~/GitHub/Dimensions/Aim3/Mothur/INPond_Initial.RData")
```

## Overview
We explored primary geographical patterns: the taxa-area relationship (TAR), the phylogenetic diversity-area relationship, and the distance-decay relationship in taxonomic and phylogenetic community similarity.

## Study area
We analyzed environmental and bacterial community data from a survey of shallow ponds found east of Bloomington, IN.
These ponds were constructed in the 1940s as wildlife refuge ponds, and are scattered throughout Brown County State Park, Yellowood State Forest, and Hoosier National Forest.
In the summer of 2013, we visited approximately 50 of these ponds and recorded their geographic locations.
We sampled aspects of water chemistry, physical properties, and bacterial community composition.

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load Environmental and Geographical Data
env <- read.table("~/GitHub/Dimensions/Aim3/DATA/EnvData/20130801_PondDataMod.csv", sep = ",", header = TRUE)
lats <- as.numeric(env[, 3]) # latitudes (north and south)
lons <- as.numeric(env[, 4]) # longitudes (east and west)
```

## Environmental data
We measured 19 environmental and geographic variables.
These included elevation (m), geographical coordinates (lat-long; data: WGS84), temperature (C), Diameter(m), Depth(m), redox potential (ORP), specific conductivity or SpC (uS/cm), dissolved Oxygen (mg/L), total dissolved solids (g/L), salinity (p.s.u.=ppm), color - measured at absorbance = 660; an estimate of carbon in the water sample, chlorophyll a (ug/ml), dissolved organic carbon (mg/L), dissolved organic nitrogen (mg/L), and total phosphorus (ug/L).

## Microbial community data
In addition to measuring a suite of geographic and environmental variables, we characterized the diversity of bacteria in the ponds using molecular-based approaches.
Specifically, we amplified the 16S rRNA gene (i.e., "DNA") and 16S rRNA transcripts (i.e., "RNA") of bacteria using barcoded primers on the Illumina MiSeq platform.
We then used a `mothur` pipeline to quality-trim our data set and assign sequences to operational taxonomic units (OTU).

For each pond, we used the observed taxonomic richness (S), total number of gene reads (N), and number of gene reads per OTU (Ni) to estimate Shannon's diversity index (H), and Simpson's evenness (D/S). 
We should estimate a handful of diversity and evenness metrics, as well conduct richness estimation for each site (Chao1, ACE, rarefaction, jacknife).
These will provide basic diversity-related variables to explore with respect to geograpy and environmental conditions.

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Select DNA Data: Use the `grep()` Command and Rename with `gsub()`
# The active portion, based on cDNA 
active.comm <- Pond97[grep("*-cDNA", rownames(Pond97)), ]
rownames(active.comm) <- gsub("\\-cDNA", "", rownames(active.comm))
rownames(active.comm) <- gsub("\\_", "", rownames(active.comm))

# The community without respect to active or not, 16S rRNA gene sequences
all.comm <- Pond97[grep("*-DNA", rownames(Pond97)), ]
rownames(all.comm) <- gsub("\\-DNA", "", rownames(all.comm))
rownames(all.comm) <- gsub("\\_", "", rownames(all.comm))

# Remove Sites Not in the Environmental Data Set
active.comm <- active.comm[rownames(active.comm)  %in% env$Sample_ID, ]
all.comm <- all.comm[rownames(all.comm)  %in% env$Sample_ID, ]

# Remove Zero-Occurrence Taxa 
active.comm <- active.comm[ , colSums(active.comm) > 0]
all.comm <- all.comm[ , colSums(all.comm) > 0]

# Import Taxonomy Data Using `read.tax()` from Source Code
tax <- read.tax(taxonomy = "~/GitHub/Dimensions/Aim3/Mothur/INPonds.bac.final.0.03.taxonomy")
```

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
#### A function to generate observed richness
S.obs <- function(x = ""){ rowSums(x > 0) * 1}
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
#### For each site:
# N equals numbers of reads
env$active.N <- as.vector(rowSums(active.comm)) 
env$all.N <- as.vector(rowSums(all.comm))

# S equals the number of non-zero abundances
env$active.S <- S.obs(active.comm)
env$all.S <- S.obs(all.comm)

# Diversity is Shannon's
env$active.H <- as.vector(diversity(active.comm, index = "shannon"))
env$all.H <- as.vector(diversity(all.comm, index = "shannon"))

# Evenness is Simpsons; divide Simpson's Diversity by S
env$active.De <- as.vector(diversity(active.comm, index = "invsimpson")/env$active.S)
env$all.De <- as.vector(diversity(all.comm, index = "invsimpson")/env$all.S)
```

## Primary geographic patterns
We examined three taxa-level geographic patterns: Distanct-decay (DD), Taxa-area relationship (TAR), and the specific spatial abundance distribution (SSAD).
While the DD and TAR have been more or less frequently studied in microbial ecology and microbial biogeography, the SSAD has been mainly, if not entirely examined in studies of macroscopic plants and animals.

### 1.) Distance Decay, taxonomic and phylogenetic
**Tobler's first law of geography** states that "Everything is related to everything else, but near things are more related than distant things" (Tobler 1970).
This law is a formulation of the concept of spatial autocorrelation.
In short, spatial autocorrelation is the degree to which spatial variables are either clustered in space (positive autocorrelation) or over-dispersed (negative autocorrelation).

The distance-decay relationship is a primary biogeographic pattern of spatial autocorrelation, and captures the rate of decreasing similarity with increasing distance. 
This pattern addresses whether communities close to one another are more similar than communities that are farther away.
The distance-decay pattern can also be used to address whether near environments have greater similarity than far ones.
We looked at decay in both taxonomic level compositional similarity via bray-curtis (should also do for Sorensens) and phylogenetic distance via unifrac distance.


# RESULTS: Distance-Decay

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
plot.new()
#par(mfrow=c(2, 2))

# Geographic Distances (Kilometers) Among Ponds
long.lat <- as.matrix(cbind(env$long, env$lat))
coord.dist <- earth.dist(long.lat, dist = TRUE)

# Taxonomic Distances Among Ponds (Bray-Curits)
active.bray.curtis.dist <- 1 - vegdist(active.comm) 
all.bray.curtis.dist <- 1 - vegdist(all.comm)

# Transform All Distances Into List Format:
active.bray.curtis.dist.ls <- liste(active.bray.curtis.dist, entry = "bray.curtis")
all.bray.curtis.dist.ls <- liste(all.bray.curtis.dist, entry = "bray.curtis")
coord.dist.ls <- liste(coord.dist, entry = "geo.dist")

# Create a Data Frame from the Lists of Distances
df <- data.frame(coord.dist.ls, active.bray.curtis.dist.ls[, 3],
          all.bray.curtis.dist.ls[, 3])
          
names(df)[4:5] <- c("active.bray.curtis", "all.bray.curtis") 
attach(df)

# Now, let's plot the DD relationships:

# Set Initial Plot Parameters
par(mfrow=c(1, 2))#, mar = c(1, 5, 2, 1) + 0.1, oma = c(2, 0, 0, 0))

# Make Plot for Taxonomic DD
plot(coord.dist, active.bray.curtis, xlab = "", xaxt = "n", las = 1, ylim = c(0.1, 0.9),
     ylab="Bray-Curtis Similarity", 
     main = "Distance Decay, Active taxa", col = "SteelBlue")

# Regression for Taxonomic DD 
DD.reg.bc <- lm(active.bray.curtis ~ geo.dist)
summary(DD.reg.bc)
abline(DD.reg.bc , col = "red4")

# Make Plot for Taxonomic DD
plot(coord.dist, all.bray.curtis, xlab = "", xaxt = "n", las = 1, ylim = c(0.1, 0.9),
     ylab="Bray-Curtis Similarity", 
     main = "Distance Decay, All taxa", col = "SteelBlue")

# Regression for Taxonomic DD 
DD.reg.bc <- lm(all.bray.curtis ~ geo.dist)
summary(DD.reg.bc)
abline(DD.reg.bc , col = "red4")

# Add X-Axis Label to Plot
mtext("Geographic Distance, km", side = 1, adj = 0.55, 
      line = 0.5, outer = TRUE)
```

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
diffslope(geo.dist, active.bray.curtis, geo.dist, all.bray.curtis)
```

### 2.) Species- or taxa- area relationship (SAR)
The species-area relationshp describes the rate at which species are discovred with increasing area. 
The SAR one of ecology's oldest and most intensively studied patterns.
Arrhenius (1921) first described the general form of the *species-area relationship (SAR)* as a power-law: $S = cA^{z}$ where S is species richnness and A is area.
Arrhenius's formula predicts a rate of increase in richness that is approximately linear in log-log space. That is, $log(S) = c + zlog(A)$, where z is the scaling exponent.

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# A function to generate the species-area relationship by
# Random Accumulating Sites

SAR.rand.accum <- function(com){
  Alist <- c()
  Slist <- c()
  
  num.ponds <- c(1,2,4,6,8,12,16,24,32,42,51)
  for (i in num.ponds) {   
    areas <- c() # hold iterated area values 
    Ss <- c() # hold iterated S values
    
    for(j in 1:5){
      pond.sample <- sample(51, replace = FALSE, size = i) 
      area <- 0
      cum.abs <- vector(length = length(com[1, ]))
  
      for (k in pond.sample) { # Loop through each randomly drawn pond
        area <- area + pond.areas[k] # aggregating area
        cum.abs <- cum.abs + com[k, ]
        } # End random pond samples loop
      
      Ss <- c(Ss, length(cum.abs[cum.abs > 0]))
      areas <- c(areas, area)
      }
    
    Alist <- rbind(Alist, mean(areas))
    Slist <- rbind(Slist, mean(Ss))
    #print(c(mean(areas), mean(Ss)))
    }
  
  return(cbind(log10(Alist), log10(Slist)))
  }

```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# A function to generate the species-area relationship by
# accumulating area according to distance

SAR.accum.dist <- function(com){
  Alist <- c()
  Slist <- c()
  num.ponds <- c(1,2,4,6,8,12,16,24,32,42,51)
  
  for (i in num.ponds) {   
    areas <- c() # hold iterated area values 
    Ss <- c() # hold iterated S values
    
    for(j in 1:5){
      pondID <- sample(51, size = 1)
      Area <- as.numeric(pond.areas[pondID]) # aggregating area
      cum.abs <- com[pondID, ]
      used <- c()
      
      for (k in 2:i) { # Loop through ponds
        sdata <- subset(coord.dist.ls, FALSE == is.element(NBX, used) & FALSE == is.element(NBY, used))
        sdata <- subset(sdata, NBX == pondID | NBY == pondID)
        sdata <- subset(sdata, geo.dist == min(sdata[, 3]))
            
        if (dim(sdata)[1] > 1) {
          x <- sample(dim(sdata)[1], size=1)
          sdata <- sdata[x,]
          }
            
        sdata <- t(as.matrix(as.numeric(as.matrix(sdata))))
        used <- c(used, as.integer(pondID))
        Area <- Area + as.numeric(pond.areas[pondID]) # aggregating area
        cum.abs <- cum.abs + com[pondID, ]
            
        if (sdata[1] - pondID == 0) {
          pondID <- sdata[2]
        } else {
          pondID <- sdata[1]
        }
      }
      Ss <- c(Ss, length(cum.abs[cum.abs > 0]))
      areas <- c(areas, Area)
    }
    # End random pond samples loop  
    Alist <- rbind(Alist, mean(areas))
    Slist <- rbind(Slist, mean(Ss))
    #print(c(mean(areas), mean(Ss)))
  }
  return(cbind(log10(Alist), log10(Slist)))
}
```


# RESULTS: Taxa-area relationship

```{r, echo=FALSE, message = FALSE, warning = FALSE}
plot.new()
par(mfrow=c(1, 1))
pond.areas <- as.vector(pi * (env$Diameter/2)^2) # Find areas of all 51 ponds

sar <- SAR.accum.dist(all.comm)
sar <- as.data.frame(sar)
plot(sar, xlab = "log(Area)", ylab = "log(Richness)",
  main = "Taxa-Area Relationship
  aggregating area by distance", col = "SteelBlue")

OLS <- lm(sar$V2 ~ sar$V1)
abline(OLS, col = "SteelBlue", lw = 2)
slope <- round(coefficients(OLS)[2], 3)
  legend("bottomright", legend = paste("slope(All) =", slope),
         bty = "n", lw = 2, col = "SteelBlue")

sar <- SAR.accum.dist(active.comm)
sar <- as.data.frame(sar)
points(sar, xlab = "log(Area)", ylab = "log(Richness)", 
  main = "Species-Area Relationship (Active)", col = "red")

OLS <- lm(sar$V2 ~ sar$V1)
abline(OLS, col = "red", lw = 2)
slope <- round(coefficients(OLS)[2], 3)
  legend("topleft", legend = paste("slope(Active) =", slope), 
         bty = "n", lw = 2, col = "red")

par(mfrow=c(1, 1))

sar <- SAR.rand.accum(all.comm)
sar <- as.data.frame(sar)
plot(sar, xlab = "log(Area)", ylab = "log(Richness)", 
  main = "Taxa-Area Relationship
  aggregating area at random", col = "SteelBlue")

OLS <- lm(sar$V2 ~ sar$V1)
abline(OLS, col = "SteelBlue", lw = 2)
slope <- round(coefficients(OLS)[2], 3)
  legend("bottomright", legend = paste("slope(All) =", slope), 
         bty = "n", lw = 2, col = "SteelBlue")

sar <- SAR.rand.accum(active.comm)
sar <- as.data.frame(sar)
points(sar, xlab = "log(Area)", ylab = "log(Richness)", 
  main = "Species-Area Relationship (Active)", col = "red")

OLS <- lm(sar$V2 ~ sar$V1)
abline(OLS, col = "red", lw = 2)
slope <- round(coefficients(OLS)[2], 3)
  legend("topleft", legend = paste("slope(Active) =", slope), 
         bty = "n", lw = 2, col = "red")
```



Let's load and process the phylogenetic data:

```{r}
# Import the Alignment File {seqinr}
ponds.cons <- read.alignment(file = "./Mothur/INPonds.bac.final.0.05.fasta", 
                             format = "fasta")  

# Rename OTUs in the FASTA File
ponds.cons$nam <- gsub("\\|.*$", "", gsub("^.*?\t", "", ponds.cons$nam))

# Import the Outgroup FASTA File {seqinr}
outgroup <- read.alignment(file = "./DATA/EnvData/methanosarcina.fasta", format = "fasta")

# Convert Alignment File to DNAbin Object {ape}
DNAbin <- rbind(as.DNAbin(outgroup), as.DNAbin(ponds.cons))

# Visusalize Sequence Alignment {ape}
image.DNAbin(DNAbin, show.labels=T, cex.lab = 0.05, las = 1) 

# Create Distance Matrix with the Jukes Cantor "JC" Model {ape}
seq.dist.jc <- dist.dna(DNAbin, model = "JC", pairwise.deletion = FALSE)

# Use Neighbor Joining Algorithm to Construct a Full Tree (DNA and RNA sequences) {ape}
phy.all <- bionj(seq.dist.jc)

# Drop Tips of Zero-Occurrence OTUs (Removes Taxa Only Found via RNA Sequencing) {ape}
phy <- drop.tip(phy.all, phy.all$tip.label[!phy.all$tip.label %in% 
                                             c(colnames(comm), "Methanosarcina")])

# Identify Outgroup Sequence
outgroup <- match("Methanosarcina", phy$tip.label)

# Root the Tree {ape}
phy <- root(phy, outgroup, resolve.root = TRUE)

# Plot the Rooted Tree {ape}
par(mar = c(1, 1, 2, 1) + 0.1)
plot.phylo(phy, main = "Neighbor Joining Tree", "phylogram", show.tip.label = FALSE, 
           use.edge.length = FALSE, direction = "right", cex = 0.6, label.offset = 1)
```

## PHYLOGENETIC ALPHA DIVERSITY

### A.  Faith's Phylogenetic Diversity (PD)
In 1992, Daniel Faith developed a diversity metric called Faith's PD (http://goo.gl/wM08Oy).
The metric sums the branch lengths for each species found in a sample from the root to the tip of a the phylogenetic tree.
The value of the metric captures the evolutionary history of an assemblage. 
Higher PD values indicate that an assemblage contains more evolutionarily divergent taxa, while lower PD values indicate that an assemblage contains taxa with a more restricted evolutionary history. 

Faith's PD can be implemented in R using the `pd()` function in the `picante` package. 
A phylogenetic tree containing the species pool is required. 
In addition to returning Faith's PD, the `pd()` function also returns species richness (SR).
SR is the same as observed richness ($S_{obs}), which we covered in the $\alpha$ diversity module. 

```{r}
# Calculate PD and SR {picante}
pd <- pd(comm, phy, include.root = FALSE)
```

Let's compare PD estimates with SR of our samples.
We'll natural-log-transform our data so that the slope of the relationship gives us a power-law exponent, which describes how PD scales with SR.

```{r}
# Biplot of SR and PD
par(mar = c(5, 5, 4, 1) + 0.1)

plot(log(pd$SR), log(pd$PD), 
     pch = 20, col = "red", las = 1,
     xlab = "ln(SR)", ylab = "ln(PD)", cex.main = 1,
     main="Phylodiversity (PD) vs. Taxonomic richness (SR)")

fit <- lm('log(pd$PD) ~ log(pd$SR)')
abline(fit, col = "red", lw = 2)
exponent <- round(coefficients(fit)[2], 2)
legend("topleft", legend=paste("Scaling exponent = ", exponent, sep = ""), 
       bty = "n", lw = 2, col = "red") 
```


#### i.  Randomizations and Null Models
Randomizations are a way of resampling data to assess whether or not observed patterns are different from a null expectation. 
A number of the functions in `picante` allow us to specify different null models as an argument. 
These null models can control for features such as species richness, species occurrence frequencey, and the diversity of the regional species pool. 
We will use some of these models for assesing the degree to which phylogenetic measures of $\alpha$ diversity deviate from null expecations. 
The following table describes some of the null models that are available to us when using `picante`:

\begin{center}
\hyphenpenalty 10000
\exhyphenpenalty 10000
\begin{tabular}{ m{5cm}  m{10cm} }
  \textbf{Null Model} & \textbf{Description} \\
  \hline \hline \\ [-1.5ex]
  \textbf{taxa.labels} & 
  Shuffles taxa labels across tips of phylogeny (across all taxa included in phylogeny) \\
  \\ [-1.5ex]
  \textbf{richness} & 
  Randomizes community data matrix abundances within samples (maintains sample species richness) \\
  \\ [-1.5ex]  
  \textbf{frequency} & 
  Randomizes community data matrix abundances within species (maintains species occurence frequency) \\
  \\ [-1.5ex]   
  \textbf{sample.pool} & 
  Randomizes community data matrix by drawing species from pool of species occurring in at least one community (sample pool) with equal probability \\
  \\ [-1.5ex]  
    \textbf{phylogeny.pool} & 
  Randomize community data matrix by drawing species from pool of species occurring in at least one community (sample pool) with equal probability \\
  \\ [-1.5ex]
  \textbf{independentswap} & 
  Randomizes community data matrix with the independent swap algorithm (Gotelli 2000) maintaining species occurrence frequency and sample species richness \\
  \\ [-1.5ex]
  \textbf{trialswap} & 
  Randomizes community data matrix with the trial-swap algorithm (Miklos \& Podani 2004) maintaining species occurrence frequency and sample species richness \\
  \\ [-1.5ex]
    \hline
\end{tabular}
\end{center}

Now, we are going to use the `ses.pd()` function in `picante`.
This function estimates the standardized effect size ("ses") using the following equation: `ses.pd` = (`pd.obs` - `pd.rand.mean`) / `pd.rand.sd`, where `pd.obs` is the observed PD, `pd.rand.mean` is the mean of the PD values generated via randomization under a null model, and `pd.rand.sd` is the standard deviation of the PD values generated via randomization under a null model (see table above). 
Given the size of both our site-by-species matrix and the phylogenetic tree, the randomizaiton process is computationally intensive. 
Therfore, we are only going to run the `ses.pd` function for two ponds with a limited number of randomizations (i.e., "runs" argument).

```{r}
# Estimate Standardized Effect Size of PD via Randomization {picante}
ses.pd <- ses.pd(comm[1:2,], phy, null.model = "richness", runs = 25, 
                 include.root = FALSE)
```


### Phylogenetic Dispersion Within a Sample
Another way to assess phylogenetic $\alpha$-diversity is to look at dispersion within a sample. 
In the following section we will introduce two commonly used metrics --- the Net Relatedness Index (NRI) and the Nearest Taxon Index --- to quantify the degree to which closely related taxa co-occur.
We will use randomization procedures to test whether species are phylogenetically clustered or overdispersed. 

#### Phylogenetic Resemblance Matrix 
Before estimating dispersion metrics, we need to create a phylogenetic resemblance matrix.
This type of matrix is nearly identical to the resemblance matrix introduced in the $\beta$-diversity module.
The only difference is that the phylogenetic resemblance matrix contains distances between taxa in a tree, whereas the community resemblance matrix contains distances among sites. 
The elements in phylogenetic resemblance matrix are calculated as the pairwise branch-length distances between tips (i.e., taxa) on a phylogenetic tree. 
The phylogenetic resemblance matrix is sometimes referred to as the phylogenetic variance-covariance matrix. 
We will use the `cophenetic.phylo()` function in `picante` to calculate the phylogenetic resemblance matrix.

```{r}
# Create a Phylogenetic Distance Matrix {picante}
phydist <- cophenetic.phylo(phy)
```

### Net Relatedness Index (NRI)
One common way to test for phylogenetic clustering and overdispersion is to use the Net Relatedness Index (NRI).
NRI is based on the mean phylogenetic distance (MPD).
MPD is the mean phylogenetic distance from pairwise branch lengths in a sample.
With this information in hand, NRI is expressed as: - (`mpd.obs` - `mpd.rand.mean`) / `mpd.rand.sd` where `mpd.obs` is the observed MPD, `mpd.rand.mean` is the mean of the MPD values generated via randomization under a null model, and `mpd.rand.sd` is the standard deviation of the MPD values generated via randomization under a null model. 

Negative NRI values indicate that a sample is phylogenetically overdispersed; that is, taxa are less related to one another than expected under the null model.
Positive NRI values indicate that a sample is phylogentically underdispersed, or clustered, such that taxa are more closely related to one another than expected under the null model. 

As with Faith's PD, the radomization procedures are computationally intensive, so we are only going to peform a relatively small number of "runs".

```{r}
# Estimate Standardized Effect Size of NRI via Randomization {picante} 
ses.mpd <- ses.mpd(comm, phydist, null.model = "taxa.labels", 
                   abundance.weighted = FALSE, runs = 25) 

# Calculate NRI
NRI <- as.matrix(-1 * ((ses.mpd[,2] - ses.mpd[,3]) / ses.mpd[,4]))
rownames(NRI) <- row.names(ses.mpd)
colnames(NRI) <- "NRI"
```

### Nearest Taxon Index (NTI)
Another way to test for phylogenetic clustering and oversdispersion in a sample is to use the Nearest Taxon Index (NTI).
This index is mathemically similar to NRI, but uses the mean nearest phylogenetic neighbor distance (MNND) instead of MPD. 
MNND is the mean phylogenetic distance between all taxa in a sample and their phylogenetically closest neighbor.
As a result, NTI tends to emphasize terminal clustering, independent of deep level clustering (Webb et al. 2002; http://goo.gl/WikgWE).
Just like NRI, we peform randomizations and use this information to estimate the standardized effect size. 
Negative NTI values indicate phylogenetic overdispersion and positive NTI values indicate phylogenetic clustering. 

```{r}
# Estimate Standardized Effect Size of NRI via Randomization {picante} 
ses.mntd <- ses.mntd(comm, phydist, null.model = "taxa.labels", 
                     abundance.weighted = FALSE, runs = 25)

# Calculate NTI
NTI <- as.matrix(-1 * ((ses.mntd[,2] - ses.mntd[,3]) / ses.mntd[,4]))
rownames(NTI) <- row.names(ses.mntd)
colnames(NTI) <- "NTI"
```

In the NRI and NTI examples above, the arguments "abundance.weighted = FALSE" means that the indices were calculated using presence-absence data.
Modify and rerun the code so that NRI and NTI are caculated using abundance data. 
How does this affect the interpretation of NRI and NTI?


## 5) PHYLOGENETIC BETA DIVERSITY 

### A. Phylogenetically Based Community Resemblance Matrix
As you may recall, in order to quantify $\beta$-diversity for more than two samples, one needs to create a resemblance matrix. 
When quantifying taxonomic $\beta$-diversity, we calculated the pairwise **similarity** or **dissimiliarity** for all samples in a site-by-species matrix using metrics such as the Sørensen index or the Bray-Curtis index. 
More or less, we need to go through the same process for quantifying phylogenetic $\beta$-diversity.
Instead of making the resemblance matrix based on incidence or abundance of taxa, we are going to incorporate information about the phylogenetic relationships among taxa. 
Similar to other measures of $\beta$-diversity, there are numerous ways to calculate the phylogenetic distances in the community resemblance matrix.
Here, we will explore two: Mean Pairwise Distance and UniFrac distance.

\begin{center}
\hyphenpenalty 10000
\exhyphenpenalty 10000
\begin{tabular}{ m{5cm}  m{10cm} }
  \textbf{Index} & \textbf{Description} \\
  \hline \hline \\ [-1.5ex]
  \textbf{Mean Pairwise Distance} & 
  Distance between two samples calculated as the mean phylogenetic distance between pairs of taxa \\
  \\ [-1.5ex]
  \textbf{UniFrac} & 
  Distance between two samples calculated as $\Sigma_{unshared}$ / $\Sigma_{total}$, where $\Sigma_{unshared}$ is the sum of unshared branch lengths between samples and $\Sigma_{total}$ is the total (shared and unshared) branch lengths in a rooted tree \\
  \\ [-1.5ex]
  \hline
\end{tabular}
\end{center}

Let's calculate the phylogenetically based community resemblance matrices for our pond data set. 
```{r, results = 'hide', message = FALSE, warning = FALSE}
# Mean Pairwise Distance
dist.mp <- comdist(comm, phydist)

# UniFrac Distance (Note: This Takes a Few Minutes; Be Patient)
dist.uf <- unifrac(comm, phy)
```

Now, let's compare the Mean Pair Distance and UniFrac distance matrices. 

```{r}
par(mar = c(5, 5, 2, 1) + 0.1)
plot(dist.mp, dist.uf, 
     pch = 20, col = "red", las = 1, asp = 1, xlim = c(0.15, 0.5), ylim = c(0.15, 0.5),
     xlab = "Mean Pair Distance", ylab = "UniFrac Distance") 
abline(b = 1, a = 0, lty = 2)
text(0.5, 0.47, "1:1")
```


### Visualizing Phylogenetic Beta Diversity
Now that we have our phylogenetically based community resemblance matrix, we can visualize phyogenetic diversity among samples using the same techniques that we used in the $\beta$-diversity module.
As an example, we will use ordination, but any of the other $\beta$-diversity viualization techniques would also work (e.g., heatmaps and cluster analysis). 
Specifically, we will use the `cmdscale()` function to conduct a Principal Coordinates Analysis (PCoA) using the UniFrac distance matrix.
Additionally, we will calculate the amount of explained variation for each phylogenetically informed PCoA axis.

```{r}
pond.pcoa <- cmdscale(dist.uf, eig = T, k = 3)

explainvar1 <- round(pond.pcoa$eig[1] / sum(pond.pcoa$eig), 3) * 100
explainvar2 <- round(pond.pcoa$eig[2] / sum(pond.pcoa$eig), 3) * 100
explainvar3 <- round(pond.pcoa$eig[3] / sum(pond.pcoa$eig), 3) * 100
sum.eig <- sum(explainvar1, explainvar2, explainvar3)
```

Now that we have calculated our PCoA, we can plot the results.
Remember, you should check the eigenvalues to determine your confidence in the data reduction approach.

```{r pcoalayer1, eval = TRUE, fig.width=6, fig.height=4}
# Define Plot Parameters
par(mar = c(5, 5, 1, 2) + 0.1)

# Initiate Plot
plot(pond.pcoa$points[ ,1], pond.pcoa$points[ ,2],
     xlim = c(-0.2, 0.2), ylim = c(-0.3, 0.15),
     xlab = paste("PCoA 1 (", explainvar1, "%)", sep = ""),
     ylab = paste("PCoA 2 (", explainvar2, "%)", sep = ""),
     pch = 16, cex = 2.0, type = "n", cex.lab = 1.5, cex.axis = 1.2, axes = FALSE)

# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
abline(h = 0, v = 0, lty = 3)
box(lwd = 2)

# Add Points & Labels
points(pond.pcoa$points[ ,1], pond.pcoa$points[ ,2],
       pch = 19, cex = 3, bg = "gray", col = "gray")
text(pond.pcoa$points[ ,1], pond.pcoa$points[ ,2], 
     labels = row.names(pond.pcoa$points))
```


#### Categorical Approach: Watershed Effect  
The ponds that we sampled in southern Indiana were located in three distinct watersheds (BCSP, HNF, and YSF). 
For many organisms, watershed boundaries represent major dispersal barriers, which may influence the phylogenetic distribution of species. 
Here, we will test for this watershed affect using the Permutational Multivariate Analysis of Variance (PERMANOVA) test that we learned about in the $\beta$-diversity module. 

```{r, results = 'hide'}
# Define Environmental Category
water.shed <- env$Location

# Run PERMANOVA with `adonis()` Function {vegan}
adonis(dist.uf ~ water.shed, permutations = 999)
```

#### ii. Continuous Approach: Environmental Gradients  
In the Geographic Ecology module, we demonstrated that there was substantial variation in enviornmental variables that are known to influence the structure and function of microbial communities.
In the following section, we will revist two methods that are used to test for relationships among multivariate environmental and biological data. 

First, let's define the environmental data and then create an enviornmental distance matrix:

```{r}
# Define Environmental Variables
envs <- env[, 5:19]

# Distance Matrix for Environmental Variables
env.dist <- vegdist(scale(envs), method = "euclid")
```

Second, let's conduct a Mantel's test to see whether UniFrac distance is correlated with environmental variation:

```{r, results = 'hide'}
# Conduct Mantel Test {vegan}
mantel(dist.uf, env.dist) 
```

Last, let's conduct a Canonical Correspondence Analysis (CCA). 
You will recall that this constrained ordination technique allows one to test for the effects of an explanatory matrix (e.g., environmental data) on a response matrix (e.g., phylogenetic distance matrix). 

```{r, results = 'hide'}
# Conduct CCA {vegan}
ponds.cca <- vegan::cca(dist.uf ~ scale(envs))

# Permutation Tests: Axes and Env Variables
anova(ponds.cca, by = "axis")
ponds.fit <- envfit(ponds.cca, envs, perm = 999)
ponds.fit

# Calculate Explained Variation
cca.explainvar1 <- round(ponds.cca$CCA$eig[1] / 
                         sum(c(ponds.cca$CCA$eig, ponds.cca$CA$eig)), 3) * 100
cca.explainvar2 <- round(ponds.cca$CCA$eig[2] / 
                         sum(c(ponds.cca$CCA$eig, ponds.cca$CA$eig)), 3) * 100

# Define Plot Parameters
par(mar = c(5, 5, 4, 4) + 0.1)

# Initiate Plot
plot(scores(ponds.cca, display = "wa"), xlim = c(-3, 3), ylim = c(-3.5, 3),
     xlab = paste("CCA 1 (", cca.explainvar1, "%)", sep = ""),
     ylab = paste("CCA 2 (", cca.explainvar2, "%)", sep = ""),
     pch = 16, cex = 2.0, type = "n", cex.lab = 1.5, cex.axis = 1.2, axes = FALSE)

# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
abline(h = 0, v = 0, lty = 3)
box(lwd = 2)

# Add Points & Labels
points(scores(ponds.cca, display = "wa"),
       pch = 19, cex = 3, bg = "gray", col = "gray")
text(scores(ponds.cca, display = "wa"), 
    labels = row.names(scores(ponds.cca, display = "wa")), cex = 0.5)

# Add Environmental Vectors
vectors <- scores(ponds.cca, display = "bp")
row.names(vectors) <- c("Elev", "Dia", "Dep", "Vol", "ORP", "Temp", "SpC", "DO", 
                        "TDS", "Sal", "pH", "Col", "Chla", "DOC", "DON")
arrows(0, 0, vectors[,1] * 2, vectors[, 2] * 2, 
       lwd = 2, lty = 1, length = 0.2, col = "red")
text(vectors[,1] * 2, vectors[, 2] * 2, pos = 3, 
     labels = row.names(vectors))
axis(side = 3, lwd.ticks = 2, cex.axis = 1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 1])) * 2, labels = pretty(range(vectors[, 1])))
axis(side = 4, lwd.ticks = 2, cex.axis = 1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 2])) * 2, labels = pretty(range(vectors[, 2])))
```



## GEOGRAPHICAL PHYLOGENETIC COMMUNITY ECOLOGY

### Phylogenetic Distance-Decay (PDD)
You will recall from the Geographical Ecology module, that the distance decay (DD) relationship reflects the spatial autocorrelation of community similarity.
That is, geographically near communities should be more similar than geographically distant communities.
Here, we will test to what degree spatial autocorrelation can also affect phylogenetic DD.  

First, we will calcualte distances for geographic data, taxonomic data, and phylogenetic data among all unique pair-wise combinations of ponds.

```{r}
# Geographic Distances (Kilometers) Among Ponds
long.lat <- as.matrix(cbind(env$long, env$lat))
coord.dist <- earth.dist(long.lat, dist = TRUE)

# Taxonomic Distances Among Ponds (Bray-Curits)
bray.curtis.dist <- 1 - vegdist(comm) 

# Phylogenetic Distances Among Ponds
unifrac.dist <- 1 - dist.uf 
  
# Transform All Distances Into List Format:
unifrac.dist.ls <- liste(unifrac.dist, entry = "unifrac")
bray.curtis.dist.ls <- liste(bray.curtis.dist, entry = "bray.curtis")
coord.dist.ls <- liste(coord.dist, entry = "geo.dist")

# Create a Data Frame from the Lists of Distances
df <- data.frame(coord.dist.ls, bray.curtis.dist.ls[, 3], unifrac.dist.ls[, 3])
names(df)[4:5] <- c("bray.curtis", "unifrac")
attach(df)
```

Now, let's plot the DD relationships:

```{r, results = 'hide'}
# Set Initial Plot Parameters
par(mfrow=c(2, 1), mar = c(1, 5, 2, 1) + 0.1, oma = c(2, 0, 0, 0))

# Make Plot for Taxonomic DD
plot(coord.dist, bray.curtis, xlab = "", xaxt = "n", las = 1, ylim = c(0.1, 0.9),
     ylab="Bray-Curtis Similarity", 
     main = "Distance Decay", col = "SteelBlue")

# Regression for Taxonomic DD 
DD.reg.bc <- lm(bray.curtis ~ geo.dist)
summary(DD.reg.bc)
abline(DD.reg.bc , col = "red4")

# New Plot Parameters
par(mar = c(2, 5, 1, 1) + 0.1)

# Make Plot for Phylogenetic DD
plot(coord.dist, unifrac, xlab = "", las = 1, ylim = c(0.1, 0.9),
     ylab = "Unifrac Similarity", col = "darkorchid4")

# Regression for Phylogenetic DD
DD.reg.uni <- lm(unifrac ~ coord.dist)
summary(DD.reg.uni)
abline(DD.reg.uni, col = "red4")

# Add X-Axis Label to Plot
mtext("Geographic Distance, km", side = 1, adj = 0.55, 
      line = 0.5, outer = TRUE)
```

Finally, let's test whether the slopes for taxonomic and phylogenetic DD are significantly different from one another.

```{r, results = 'hide'}
diffslope(geo.dist, unifrac, geo.dist, bray.curtis)
```


### Phlogenetic diversity-area relationship (PDAR)

Helmus and Ives (2012) developed methods to study how phylogenetic diversity changes with increasing area.
This phylogenetic diversity-area relationship (PDAR) is analogous to the species-area relationship (SAR) that we learned about in the Geographical Ecology module.
You will recall that the SAR is a cumulative relationship (S = cA^z) and so, cannot be negative.
In fact, the shallowest the SAR can get is to have a slope of 0.0, meaning that all species are found in all samples.
Below is a plot of the taxonomically based SAR for the pond data. 

In contrast to SAR, phylogenetic divesity-area relationships can increase or decrease with area. 
This provoked Helmus and Ives (2012) to develop theoretical expectations for the effects that ecological and evolutionary processes might have on the shape of the PDAR.

#### i. Constructing the PDAR  
Helmus and Ives (2012) used the phylogenetic species variability (PSV) metric to quantify phylogenetic diversity.
PSV quantifies how phylogenetic relatedness decreases the variance of a hypothetical neutral trait shared by all species in a community. 
Type 'help(psv)' to learn more.

In this exercise, we will construct PDARs using the random aggregation approach of Helmus and Ives (2012).
This approach is similar to how we constructed SARs in the Geographical Ecology module.

Let's begin by writing a function to generate the PDAR.

```{r}
PDAR <- function(comm, tree){

  # Create Objects to Hold Areas and Diversity
  areas <- c() 
  diversity <- c() 
  
  # Create Vector Increasing Number of Plots by 2x
  num.plots <- c(2, 4, 8, 16, 32, 51)
  
  for (i in num.plots){   
    # Create vectors to hold areas and diversity form iterations, used for means
    areas.iter <- c() 
    diversity.iter <- c() 
    
    # Iterate 10 Times Per Sample Size
    for (j in 1:10){ 
      # Sample w/o replacement
      pond.sample <- sample(51, replace = FALSE, size = i)
      
      # Create Variable and Vector to Hold Accumulating Area and Taxa
      area <- 0 
      sites <- c() 
      
      for (k in pond.sample) {       # Loop Through Each Randomly Drawn Pond  
        area <- area + pond.areas[k] # Aggregating Area (Roughly Doubling)
        sites <- rbind(sites, comm[k, ])    # And Sites
        }
      
      # Concatenate the rea to areas.iter
      areas.iter <- c(areas.iter, area)  
      # Calculate PSV or Other Phylogenetic Alpha-Diversity Metric
      psv.vals <- psv(sites, tree, compute.var = FALSE) 
      psv <- psv.vals$PSVs[1]
      diversity.iter <- c(diversity.iter, as.numeric(psv))
      }

    diversity <- c(diversity, mean(diversity.iter)) # Let Diversity be the Mean PSV
    areas <- c(areas, mean(areas.iter))             # Let areas be the Average Area
    print(c(i, mean(diversity.iter), mean(areas.iter))) # Print As We Go
    }
  # Return Vectors of Areas (x) and Diversity (y)
  return(cbind(areas, diversity)) 
  }
```

#### ii. Evaluating the PDAR  
We will examine the relationship between phylogenetic diversity and area using both Spearman's correlation coefficient (S) and Pearson's correlation coefficient (P).
It is informative to use both because while S is computed on ranks and depicts monotonic relationships (the degree to which the relationship is continually increasing or decreasing), P is computed on the observed values and therefore depicts linear relationships.

```{r, results = 'hide'}
# Calculate Areas for Ponds: Find Areas of All 51 ponds
pond.areas <- as.vector(pi * (env$Diameter/2)^2) 

# Compute the PDAR
pdar <- PDAR(comm, phy) 
pdar <- as.data.frame(pdar)
pdar$areas <- sqrt(pdar$areas)

# Calculate Perason's Correlation Coefficient
Pearson <- cor.test(pdar$areas, pdar$diversity, method = "pearson")
P <- round(Pearson$estimate, 2)
Pp <- round(Pearson$p.value, 3)

# Calculate Spearman's Correlation Coefficient
Spearman <- cor.test(pdar$areas, pdar$diversity, method = "spearman")
S <- round(Spearman$estimate, 2)
Sp <- round(Spearman$p.value, 3)

# Plot the PDAR
par(mar = c(5, 5, 4, 2) + 0.1)
plot(pdar[, 1], pdar[, 2], xlab = "Area", ylab = "PSV", ylim = c(0, 1), 
     main = "Phylogenetic Diversity-Area Relationship", 
     col = "red", pch = 16, las = 1)

legend("topleft", legend= c(paste("Spearman Correlation = ", S, "; p = ", Sp, sep = ""), 
                            paste("Pearson Correlation = ", P, "; p = ", Pp, sep = "")), 
                            bty = "n", col = "red") 
```
