---
title: "Supplemental Figures: Compositional similarity vs. Geographic distance"
geometry: margin=2.54cm
date: "November 7, 2015"
output: pdf_document
header-includes: \usepackage{array}
---

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Retrieve and set the working directory
rm(list=ls())
getwd()
setwd("~/GitHub/Dimensions/Aim3/papers/DD")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load packages
require("sp")          # Classes and methods for handling spatial data
require("geoR")        # Methods for geostatistical analyses
require("rgdal")       # Geospatial Data Abstraction Library
require("raster")      # Methods to create a RasterLayer object
require("maptools")    # Tools for manipulating and reading geospatial data

require("picante")
require("ape")
require("seqinr")
require("vegan") # biodiversity estimators and related functions
require("fossil")
require("simba")
require("reshape")

source("~/GitHub/Dimensions/Aim3/bin/MothurTools.R")
source("~/GitHub/Dimensions/Aim3/bin/metrics.R")
load(file = "~/GitHub/Dimensions/Aim3/Mothur/INPond_Initial.RData")
```


```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load Environmental and Geographical Data
env <- read.table("~/GitHub/Dimensions/Aim3/DATA/EnvData/20130801_PondDataMod.csv", sep = ",", header = TRUE)
lats <- as.numeric(env[, 3]) # latitudes (north and south)
lons <- as.numeric(env[, 4]) # longitudes (east and west)

# Geographic Distances (Kilometers) Among Ponds
long.lat <- as.matrix(cbind(env$long, env$lat))
geo.dist <- earth.dist(long.lat, dist = TRUE)
#geo.dist <- log(geo.dist)
geo.dist[which(!is.finite(geo.dist))] = NA
geo.dist.ls <- liste(geo.dist, entry = "geo.dist")

# Select DNA Data: Use the `grep()` Command and Rename with `gsub()`
# The active portion, based on cDNA 
active.comm <- Pond97[grep("*-cDNA", rownames(Pond97)), ]
rownames(active.comm) <- gsub("\\-cDNA", "", rownames(active.comm))
rownames(active.comm) <- gsub("\\_", "", rownames(active.comm))

# The community without respect to active or not, 16S rRNA gene sequences
all.comm <- Pond97[grep("*-DNA", rownames(Pond97)), ]
rownames(all.comm) <- gsub("\\-DNA", "", rownames(all.comm))
rownames(all.comm) <- gsub("\\_", "", rownames(all.comm))

# Remove Sites Not in the Environmental Data Set
active.comm <- active.comm[rownames(active.comm)  %in% env$Sample_ID, ]
all.comm <- all.comm[rownames(all.comm)  %in% env$Sample_ID, ]

# Remove Zero-Occurrence Taxa 
active.comm <- active.comm[ , colSums(active.comm) > 0]
all.comm <- all.comm[ , colSums(all.comm) > 0]

cutoff <- 1000
all.comm <- all.comm[, 1:cutoff]
active.comm <- active.comm[, 1:cutoff]
```


```{r}
# Conduct CCA and get axis 1
# Define Environmental Matrix
# Pond environmental variables
env.dat <- as.matrix(subset(env, select = Diameter:DON))
env.chem <- env.dat
# Conduct CCA 
active.cca <- vegan::cca(active.comm ~ env.chem)
all.cca <- vegan::cca(all.comm ~ env.chem)

# Permutation Tests
#anova(active.cca, by = "axis")
active.cca.fit <- envfit(active.cca, env.chem, perm = 999)
active.cca.scores <- scores(active.cca)
active.cca.sites <- as.data.frame(scores(active.cca.scores$sites))
active.cca1.scores <- as.matrix(subset(active.cca.sites, select = CCA1))
active.cca1.dist <- vegdist(active.cca1.scores, "euclidean")

#anova(all.cca, by = "axis")
all.cca.fit <- envfit(all.cca, env.chem, perm = 999)
all.cca.scores <- scores(all.cca)
all.cca.sites <- as.data.frame(scores(all.cca.scores$sites))
all.cca1.scores <- as.matrix(subset(all.cca.sites, select = CCA1))
all.cca1.dist <- vegdist(all.cca1.scores, "euclidean")
```



```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE, fig.width=6, fig.height=6}
metric <- "jaccard"

# Taxonomic Distances Among Ponds (Bray-Curits)
active.dist <- log10(1 - vegdist(active.comm, method=metric, binary=FALSE))
all.dist <- log10(1 - vegdist(all.comm, method=metric, binary=FALSE))

# Regression for active geo
DD.active.geo <- lm(active.dist ~ geo.dist)
summary(DD.active.geo, correlation = TRUE)

# Regression for all geo
DD.all.geo <- lm(all.dist ~ geo.dist)
coeff <- summary(DD.all.geo, correlation = TRUE)  

# Regression for active env
DD.active.env <- lm(active.dist ~ active.cca1.dist)
summary(DD.active.env, correlation = TRUE)

# Regression for all env
DD.all.env <- lm(all.dist ~ all.cca1.dist)
coeff <- summary(DD.all.env, correlation = TRUE)
```



```{r}
file <- paste("~/GitHub/Dimensions/Aim3/papers/DD/DD/figs/DDRs_Fig2.png", sep="")
png(filename=file)
  
plot.new()
par(mfrow=c(2, 2), mar = c(5, 4, 3, 2) + 0.1, oma =c(0,1,3,0))

# Make Plot for all geo
slope <- round(coefficients(DD.all.geo)[2], 3)
p <- round(summary(DD.all.geo)$coefficients[8], 3)
plot(geo.dist, all.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab="Distance, km", col = "azure3", cex.lab=1.5)
abline(DD.all.geo, col = "red")
legend("bottomleft", paste("All slope=",slope," p = ",p), bty="n", cex=1.2)


# Make Plot for active geo
slope <- round(coefficients(DD.active.geo)[2], 3)
p <- round(summary(DD.active.geo)$coefficients[8],3)
plot(geo.dist, active.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab="Distance, km",col = "azure3",cex.lab=1.5)
abline(DD.active.geo, col = "red")
legend("bottomleft", paste("Active slope=",slope," p = ",p), bty="n", cex=1.2)


# Make Plot for all cca1
slope <- round(coefficients(DD.all.env)[2], 3)
p <- round(summary(DD.all.env)$coefficients[8], 3)
plot(all.cca1.dist, all.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab="CCA1", col = "azure3", cex.lab=1.5)
abline(DD.all.env, col = "red")
legend("bottomleft", paste("All: m = ",slope," p = ",p), bty="n", cex=1.2)

# Make Plot for active cca1
slope <- round(coefficients(DD.active.env)[2], 3)
p <- round(summary(DD.active.env)$coefficients[8],3)
plot(active.cca1.dist, active.dist, xaxt = "s", las = 1, 
   ylab="log(similarity)", xlab="CCA1",col = "azure3",cex.lab=1.5)
abline(DD.active.env, col = "red")
legend("bottomleft", paste("Active: m = ",slope," p = ",p), bty="n", cex=1.2)


# Add X-Axis Label to Plot
mtext("Geographic Distance, km", side = 1, adj = 0, outer = TRUE)
  
#d1 <- diffslope(geo.dist, active.dist, geo.dist, all.dist)
#d2 <- diffslope(active.cca1.dist, active.dist, all.cca1.dist, all.dist)

Mtitle <- paste('Geographic DDRs: Difference in slope =', 
                round(as.numeric(d1[2]),3), '; p =', d1[3],"\n",
                'Environmental DDRs: Difference in slope =', 
                round(as.numeric(d2[2]),3), '; p =', d2[3])

title(main=Mtitle, line=-2, outer=T, cex.main=1.5)

dev.off()
```
