---
title: "Spatial vs. temporal storage: The dual role of microbial seed
banks in driving geographical patterns of microbial diversity"

geometry: margin=2.54cm
date: "August 4, 2015"
output: pdf_document
header-includes: \usepackage{array}
---

```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Retrieve and set the working directory
rm(list=ls())
getwd()
setwd("~/GitHub/Dimensions/Aim3")
```




```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load packages
require("sp")          # Classes and methods for handling spatial data
require("geoR")        # Methods for geostatistical analyses
require("rgdal")       # Geospatial Data Abstraction Library
require("raster")      # Methods to create a RasterLayer object
require("maptools")    # Tools for manipulating and reading geospatial data

require("picante")
require("ape")
require("seqinr")
require("vegan") # biodiversity estimators and related functions
require("fossil")
require("simba")
require("reshape")

require(ggplot2)
require(reshape)
```



```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Import functions
# import code file that includes a function for reading in output files 
#from community sequencing software `mothur` (http://www.mothur.org/).
source("~/GitHub/Dimensions/Aim3/bin/MothurTools.R")
load(file = "~/GitHub/Dimensions/Aim3/Mothur/INPond_Initial.RData")
```



```{r, results = 'hide', echo=FALSE, message = FALSE, warning = FALSE}
# Load Environmental and Geographical Data
env <- read.table("~/GitHub/Dimensions/Aim3/DATA/EnvData/20130801_PondDataMod.csv",
                  sep = ",", header = TRUE)
lats <- as.numeric(env[, 3]) # latitudes (north and south)
lons <- as.numeric(env[, 4]) # longitudes (east and west)

# Geographic Distances (Kilometers) Among Ponds
long.lat <- as.matrix(cbind(env$long, env$lat))
coord.dist <- earth.dist(long.lat, dist = TRUE)
coord.dist <- log(coord.dist)
coord.dist[which(!is.finite(coord.dist))] = NA
coord.dist.ls <- liste(coord.dist, entry = "geo.dist")

# Select DNA Data: Use the `grep()` Command and Rename with `gsub()`
# The active portion, based on cDNA 
active.comm <- Pond97[grep("*-cDNA", rownames(Pond97)), ]
rownames(active.comm) <- gsub("\\-cDNA", "", rownames(active.comm))
rownames(active.comm) <- gsub("\\_", "", rownames(active.comm))

# The community without respect to active or not, 16S rRNA gene sequences
all.comm <- Pond97[grep("*-DNA", rownames(Pond97)), ]
rownames(all.comm) <- gsub("\\-DNA", "", rownames(all.comm))
rownames(all.comm) <- gsub("\\_", "", rownames(all.comm))

# Remove Sites Not in the Environmental Data Set
active.comm <- active.comm[rownames(active.comm)  %in% env$Sample_ID, ]
all.comm <- all.comm[rownames(all.comm)  %in% env$Sample_ID, ]

# Remove Zero-Occurrence Taxa 
active.comm <- active.comm[ , colSums(active.comm) > 0]
all.comm <- all.comm[ , colSums(all.comm) > 0]
```




## 1.) Variance partitioning: 
```{r, results='hide', echo=F, message=F, warning=F, fig.width=7, fig.height=4}

cutoffs <- c(15, 30, 60, 125, 250, 500, 1000, 2000, 4000, 8000, 16000, 22098)

active.env <- c()
active.geo <- c()
active.env.geo <- c()

all.env <- c()
all.geo <- c()
all.env.geo <- c()

# Geographic variables
geo.dat <- as.matrix(subset(env, select = lat:Elevation))
# Pond environmental variables
env.dat <- as.matrix(subset(env, select = Diameter:DON))

for (cutoff in cutoffs){
  print(cutoff)
  all.sub.comm <- all.comm[, 1:cutoff]
  active.sub.comm <- active.comm[, 1:cutoff]
  
  ## ALL
  #rda.geo <- rda(all.sub.comm ~ geo.dat)
  #rda.env <- rda(all.sub.comm ~ env.dat)
  
  # Two explanatory matrices -- Hellinger-transform Y
  mod <- varpart(Y=all.sub.comm, X=geo.dat, env.dat, transfo="hellinger")
  
  all.env <- c(all.env, mod[[1]][[3]][[3]][3])
  all.geo <- c(all.geo, mod[[1]][[3]][[3]][1])
  all.env.geo <- c(all.env.geo, mod[[1]][[3]][[3]][2])
  
  # Test fraction [a] using RDA:
  #geo.anova <- anova(rda.geo, step=200, perm.max=200)
  #geo.anova
  
  ## ACTIVE
  #rda.geo <- rda(active.sub.comm ~ geo.dat)
  #rda.env <- rda(active.sub.comm ~ env.dat)
  
  # Two explanatory matrices -- Hellinger-transform Y
  mod <- varpart(Y=active.sub.comm, X=geo.dat, env.dat, transfo="hellinger")
  
  active.env <- c(active.env, mod[[1]][[3]][[3]][3])
  active.geo <- c(active.geo, mod[[1]][[3]][[3]][1])
  active.env.geo <- c(active.env.geo, mod[[1]][[3]][[3]][2])
  }

# Set Initial Plot Parameters
plot.new()
par(mfrow=c(2, 2))

varpar.df <- data.frame(cutoff = cutoffs,
                 All_Env = all.env, 
                 All_Geography = all.geo,
                 All_GeoEnv = all.env.geo,
                 Active_Env = active.env, 
                 Active_Geography = active.geo,
                 Active_GeoEnv = active.env.geo)
  
varpar.df <- melt(varpar.df,  id.vars = 'cutoff', variable_name = 'series')

# plot on same grid, each series colored differently
ggplot(varpar.df, aes(cutoff, value)) +
  geom_line(aes(colour = series)) +
  labs(x= "Number of OTU's", y="Explained variance") +
  labs(title="Dependence of variance partitioning on rare taxa") +
  theme(legend.position="top") +
  guides(fill=guide_legend(title=NULL)) +
  theme(legend.title=element_blank())
```



### 2.) Constrained Ordination
Another way we can test hypotheses with continuous data is to use **constrained ordination**, which is sometimes referred to as canonical ordination.
Constrained ordination explores the relationships between two matrices: an **explanatory matrix** and a **response matrix**. 
Canonical correspondence analysis (CCA) and redundancy analysis (RDA) are two types of constrained ordination.
These techniques are based on the linear model framework and thus can be used to formally test hypotheses.
Constrained ordination works by first conducting multivariate multiple linear regression followed either by correspondence analysis (CA) with CCA or Principal Components Analysis (PCA) with RDA, while using the matrix of fitted values to obtain a constrained ordination.
A permutation test can then be used to test for overall significance. 

Here, we will use environmental data to conduct a CCA on the fish assemblages of the Doubs River. 
We will start by creating an explanatory matrix that contains water chemistry data.
We will then use the `cca()` function from the `vegan` package.
Note, we have to specify that we want the `cca` function in the `vegan` package because there are `cca` functions in both `vegan` and `ade4`!
We will then use permutation tests to evaluate the significance of our model. 
Finally, we will test the influence of each environmental variable on the constrained axes.

```{r, results = "hide"}
# Get Dominant members
cutoffs <- c(15, 30, 60, 125, 250)#, 500, 1000, 2000, 4000, 8000, 16000, 22098)
cutoff <- 30

all.sub.comm <- all.comm[, 1:cutoff]
active.sub.comm <- active.comm[, 1:cutoff]


# Define Environmental Matrix
env.chem <- env.dat

# Conduct CCA 
active.cca <- vegan::cca(active.sub.comm ~ env.chem)

# Permutation Tests
anova(active.cca, by = "axis")
cca.fit <- envfit(active.cca, env.chem, perm = 999)
cca.fit

# Calculate Explained Variation
cca.explainvar1 <- round(active.cca$CCA$eig[1] / 
                         sum(c(active.cca$CCA$eig, active.cca$CA$eig)), 3) * 100
cca.explainvar2 <- round(active.cca$CCA$eig[2] / 
                         sum(c(active.cca$CCA$eig, active.cca$CA$eig)), 3) * 100

# Define Plot Parameters
par(mar = c(5, 5, 4, 4) + 0.1)

# Initiate Plot
plot(scores(active.cca, display = "wa"), xlim = c(-3.5, 2), ylim = c(-3.2, 3.2),
     xlab = paste("CCA 1 (", cca.explainvar1, "%)", sep = ""),
     ylab = paste("CCA 2 (", cca.explainvar2, "%)", sep = ""),
     pch = 16, cex = 2.0, type = "n", cex.lab = 1.5, cex.axis = 1.2, axes = FALSE)

# Add Axes
axis(side = 1, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
axis(side = 2, labels = T, lwd.ticks = 2, cex.axis = 1.2, las = 1)
abline(h = 0, v = 0, lty = 3)
box(lwd = 2)

# Add Points & Labels
points(scores(active.cca, display = "wa"),
       pch = 19, cex = 3, bg = "gray", col = "gray")
text(scores(active.cca, display = "wa"), 
     labels = row.names(scores(active.cca, display = "wa")))

# Add Environmental Vectors
vectors <- scores(active.cca, display = "bp")
row.names(vectors) <- c("diameter", "depth", "volume", "ORP", "temp", "SpC", "DO", "TDS", "salinity", "pH", "color", "chla", "DOC", "DON")

arrows(0, 0, vectors[,1] * 2, vectors[, 2] * 2, 
       lwd = 2, lty = 1, length = 0.2, col = "red")
text(vectors[,1] * 2, vectors[, 2] * 2, pos = 3, 
     labels = row.names(vectors))
axis(side = 3, lwd.ticks=2, cex.axis=1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 1])) * 2, labels = pretty(range(vectors[, 1])))
axis(side = 4, lwd.ticks=2, cex.axis=1.2, las = 1, col = "red", lwd = 2.2,
     at = pretty(range(vectors[, 2])) * 2, labels = pretty(range(vectors[, 2])))
```

